var documenterSearchIndex = {"docs":
[{"location":"nnmm/Part1_introduction/#Part1.-introduction","page":"Part 1. Introduction","title":"Part1. introduction","text":"","category":"section"},{"location":"nnmm/Part1_introduction/#1.-Overview","page":"Part 1. Introduction","title":"1. Overview","text":"The Mixed Effect Neural Networks (NN-MM) extend linear mixed model (\"MM\") to multilayer neural networks (\"NN\") by adding one middle layer between genotype layer and phenotypes layer. Nodes in the middle layer represent intermediate traits, e.g., the known intermediate omics features such as gene expression levels can be incorporated in the middle layer. These three sequential layers form a unified network. \n\n(Image: )\n\nNN-MM allow any patterns of missing data in the middle layer, and missing data will be sampled. In below figure, for an individual, the gene expression levels of the first two genes are 0.9 and 0.1, respectively, and the gene expression level of the last gene is missing to be sampled. The missing patterns of gene expression levels can be different for different individuals.","category":"section"},{"location":"nnmm/Part1_introduction/#2.-Extend-linear-mixed-model-to-multilayer-neural-networks","page":"Part 1. Introduction","title":"2. Extend linear mixed model to multilayer neural networks","text":"Multiple independent single-trait mixed models are used to model the relationships between input layer (genotypes) and middle layer (intermediate traits). Activation functions in the neural network are used to approximate the linear/nonlinear relationships between middle layer (intermediate traits) and output layer (phenotypes). Missing values in the middle layer (intermediate traits) are sampled by Hamiltonian Monte Carlo based on the upstream genotype layer and downstream phenotype layer.\n\nDetails can be found in our publications:\n\nTianjing Zhao, Jian Zeng, and Hao Cheng. Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data, GENETICS, 2022; https://doi.org/10.1093/genetics/iyac034. \nTianjing Zhao, Rohan Fernando, and Hao Cheng. Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies, G3 Genes|Genomes|Genetics, 2021;  https://doi.org/10.1093/g3journal/jkab228","category":"section"},{"location":"nnmm/Part1_introduction/#3.-Flexibility","page":"Part 1. Introduction","title":"3. Flexibility","text":"NN-MM can fit fully-connected neural networks ((a),(b)), or partial-connected neural networks ((c),(d)). Also, the relationship between middle layer (intermediate traits) and output layer (phenotypes) can be based on activation functions ((a),(c)), or pre-defined by a user-defined function ((b),(d)).\n\n(Image: )","category":"section"},{"location":"nnmm/Part1_introduction/#4.-Multi-threaded-parallelism","page":"Part 1. Introduction","title":"4. Multi-threaded parallelism","text":"By default, multiple single-trait models will be used to model the relationships between input layer (genotypes) and middle layer (intermediate traits). Multi-threaded parallelism will be used for parallel computing. The number of threads can be checked by running Threads.nthreads() in Julia. Usually, using multiple threads will be about 3 times faster than using a single thread.\n\nThe number of execution threads is controlled by using the -t/–threads command-line argument (requires at least Julia 1.5). \n\nFor example, to start Julia with 4 threads:\n\njulia --threads 4\n\nIf you're using Juno via IDE like Atom, all threads will be loaded automatically. ","category":"section"},{"location":"manual/getstarted/#Get-Started","page":"Get Started","title":"Get Started","text":"","category":"section"},{"location":"manual/getstarted/#Installation","page":"Get Started","title":"Installation","text":"To install Julia, please go to the official Julia website. Please see platform specific instructions if you have trouble installing Julia.\n\nTo install the NNMM package, use the following command inside the Julia REPL (or IJulia Notebook):\n\nusing Pkg\nPkg.add(\"NNMM\")\n\nTo load the NNMM package:\n\nusing NNMM","category":"section"},{"location":"manual/getstarted/#Development-Version","page":"Get Started","title":"Development Version","text":"To use the latest/beta features under development:\n\nPkg.add(url=\"https://github.com/reworkhow/NNMM.jl\")","category":"section"},{"location":"manual/getstarted/#Jupyter-Notebook","page":"Get Started","title":"Jupyter Notebook","text":"If you prefer \"reproducible research\", an interactive Jupyter Notebook interface is available for Julia (and therefore NNMM). The Jupyter Notebook is an open-source web application for creating and sharing documents that contain live code, equations, visualizations and explanatory text. To install IJulia for Jupyter Notebook, please go to IJulia.","category":"section"},{"location":"manual/getstarted/#Multi-threaded-Parallelism","page":"Get Started","title":"Multi-threaded Parallelism","text":"NNMM supports multi-threaded parallelism for faster computation. The number of threads can be checked by:\n\nThreads.nthreads()\n\nTo start Julia with multiple threads (requires Julia 1.5+):\n\njulia --threads 4","category":"section"},{"location":"manual/getstarted/#Access-Documentation","page":"Get Started","title":"Access Documentation","text":"warning: Warning\nPlease load the NNMM package first.\n\nTo show basic information about NNMM in REPL or IJulia notebook, use ?NNMM and press enter.\n\nFor help on a specific function, type ? followed by its name, e.g., ?runNNMM and press enter.\n\nThe full documentation is available here.","category":"section"},{"location":"manual/getstarted/#Run-Your-Analysis","page":"Get Started","title":"Run Your Analysis","text":"There are several ways to run your analysis:","category":"section"},{"location":"manual/getstarted/#Interactive-Session-(REPL)","page":"Get Started","title":"Interactive Session (REPL)","text":"Start an interactive session by double-clicking the Julia executable or running julia from the command line:\n\njulia> using NNMM\njulia> # your analysis code here\n\nTo evaluate code written in a file script.jl in REPL:\n\njulia> include(\"script.jl\")\n\nTo exit the interactive session, type ^D (control + d) or quit().","category":"section"},{"location":"manual/getstarted/#Command-Line","page":"Get Started","title":"Command Line","text":"To run code in a file non-interactively from the command line:\n\njulia script.jl\n\nIf you want to pass arguments to your script:\n\njulia script.jl arg1 arg2\n\nwhere arguments arg1 and arg2 are passed as ARGS[1] and ARGS[2] of type String.","category":"section"},{"location":"manual/getstarted/#Jupyter-Notebook-2","page":"Get Started","title":"Jupyter Notebook","text":"To run code in Jupyter Notebook, please see IJulia.","category":"section"},{"location":"nnmm/Part2_NNMM/#Mixed-effect-neural-network:-Genotypes-Unobserved-intemediate-traits-Phenotyes","page":"Part 2. NNMM","title":"Mixed effect neural network: Genotypes -> Unobserved intemediate traits -> Phenotyes","text":"Tips: please center the phenotypes to have zero mean.","category":"section"},{"location":"nnmm/Part2_NNMM/#Example(a):-fully-connected-neural-networks,-all-intemediate-traits-are-unobserved","page":"Part 2. NNMM","title":"Example(a): fully-connected neural networks, all intemediate traits are unobserved","text":"nonlinear function (to define relationship between middle layer and phenotye): tanh (other supported activation functions: \"sigmoid\", \"relu\", \"leakyrelu\", \"linear\")\nnumber of nodes in the middle layer: 3\nBayesian model: multiple independent single-trait BayesC (to sample marker effects on intemediate traits). Note, to use multi-trait Bayesian Alphabet models, please set mega_trait=false in runMCMC() function.\nsample the unobserved intemediate traits in the middle layer: Hamiltonian Monte Carlo\n\n(Image: )\n\n# Step 1: Load packages\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets,Random\nRandom.seed!(123)\n\n# Step 2: Read data \nphenofile  = dataset(\"phenotypes.csv\") #get example data path\ngenofile   = dataset(\"genotypes.csv\")  #get example data path\n\nphenotypes = CSV.read(phenofile,DataFrame,delim = ',',header=true,missingstrings=[\"NA\"]) #read phenotypes (output layer)\ngenotypes  = get_genotypes(genofile,separator=',',method=\"BayesC\");                      #read genotypes  (input layer)\n\n\n# Step 3: Build Model Equations \nmodel_equation  =\"y1 = intercept + genotypes\"  #name of phenotypes is \"y1\" in the phenotypes data\n                                               #name of genotypes is \"genotypes\" (user-defined in the previous step)\n                                               #the single-trait mixed model used between input and each node in middle layer is: middle node = intercept + genotypes\nmodel = build_model(model_equation,\n\t\t    num_hidden_nodes=3,            #number of nodes in middle layer is 3\n\t\t    nonlinear_function=\"tanh\");    #tanh function is used to approximate relationship between middle layer and phenotype\n\n\n# Step 4: Run Analysis\nout=runMCMC(model,phenotypes,chain_length=5000); \n\n# Step 5: Check Accuruacy\nresults    = innerjoin(out[\"EBV_NonLinear\"], phenotypes, on = :ID) \naccuruacy  = cor(results[!,:EBV],results[!,:bv1])","category":"section"},{"location":"nnmm/Part2_NNMM/#Example-output-files","page":"Part 2. NNMM","title":"Example output files","text":"The i-th middle nodes will be named as \"trait name\"+\"i\". In our example, the observed trait is named \"y1\", and there are 3 middle nodes, so the middle nodes are named as \"y11\", \"y12\", and \"y13\", respectively.\n\nBelow is a list of files containing estimates and standard deviations for variables of interest. \n\nfile name description\nEBV_NonLinear.txt estimated breeding values for observed trait\nEBV_y11.txt estimated breeding values for middle node 1\nEBV_y12.txt estimated breeding values for middle node 2\nEBV_y13.txt estimated breeding values for middle node 3\ngenetic_variance.txt estimated genetic variance-covariance of all middle nodes\nheritability.txt estimated heritability of all middle nodes\nlocation_parameters.txt estimated bias of all middle nodes\nneuralnetworksbiasandweights.txt. estimated bias of phenotypes and weights between middle nodes and phenotypes\npi_genotypes.txt estimated pi of all middle nodes\nmarkereffectsgenotypes.txt estimated marker effects of all middle nodes\nresidual_variance.txt estimated residual variance-covariance for all middle nodes\n\nBelow is a list of files containing MCMC samples for variables of interest. \n\nfile name description\nMCMCsamplesEBV_NonLinear.txt MCMC samples from the posterior distribution of breeding values for phenotypes\nMCMCsamplesEBV_y11.txt MCMC samples from the posterior distribution of breeding values for middle node 1\nMCMCsamplesEBV_y12.txt MCMC samples from the posterior distribution of breeding values for middle node 2\nMCMCsamplesEBV_y13.txt MCMC samples from the posterior distribution of breeding values for middle node 3\nMCMCsamplesgenetic_variance.txt MCMC samples from the posterior distribution of genetic variance-covariance for all middle nodes\nMCMCsamplesheritability.txt MCMC samples from the posterior distribution of heritability for all middle nodes\nMCMCsamplesmarkereffectsgenotypes_y11 MCMC samples from the posterior distribution of marker effects for middle node 1\nMCMCsamplesmarkereffectsgenotypes_y12 MCMC samples from the posterior distribution of marker effects for middle node 2\nMCMCsamplesmarkereffectsgenotypes_y13 MCMC samples from the posterior distribution of marker effects for middle node 3\nMCMCsamplesmarkereffectsvariances_genotypes.txt MCMC samples from the posterior distribution of marker effect variance for all middle nodes\nMCMCsamplesneuralnetworksbiasandweights.txt. MCMC samples from the posterior distribution of bias of observed trait and weights between middle nodes and phenotypes\nMCMCsamplespi_genotypes.txt MCMC samples from the posterior distribution of pi for all middle nodes\nMCMCsamplesresidual_variance.txt MCMC samples from the posterior distribution of residual variance-covariance for all middle nodes","category":"section"},{"location":"manual/public/#Public-Functions","page":"Public Functions","title":"Public Functions","text":"Documentation for NNMM.jl's public interface. Below are functions available to general users.","category":"section"},{"location":"manual/public/#Index","page":"Public Functions","title":"Index","text":"Pages = [\"public.md\"]\nModules = [NNMM]","category":"section"},{"location":"manual/public/#Core-Types","page":"Public Functions","title":"Core Types","text":"Types for defining the NNMM network architecture.","category":"section"},{"location":"manual/public/#Main-Functions","page":"Public Functions","title":"Main Functions","text":"Primary functions for running NNMM analysis.","category":"section"},{"location":"manual/public/#Data-Reading-Functions","page":"Public Functions","title":"Data Reading Functions","text":"Functions for reading input data files.","category":"section"},{"location":"manual/public/#Post-Analysis","page":"Public Functions","title":"Post-Analysis","text":"Functions for analyzing MCMC results.","category":"section"},{"location":"manual/public/#Built-in-Datasets","page":"Public Functions","title":"Built-in Datasets","text":"","category":"section"},{"location":"manual/public/#Pedigree-Functions","page":"Public Functions","title":"Pedigree Functions","text":"","category":"section"},{"location":"manual/public/#NNMM.Layer","page":"Public Functions","title":"NNMM.Layer","text":"Layer\n\nA struct representing a layer in the Neural Network Mixed Model (NNMM).\n\nFields\n\nlayer_name::String: Name identifier for the layer\ndata_path: Path to the data file(s) - String for single file, Vector{String} for partial connection\nseparator: Column separator in the data file (default: ',')\nheader: Whether the data file has a header row (default: true)\ndata: Loaded data (initialized as empty)\nquality_control: Whether to perform quality control on genotypes (default: true)\nMAF: Minor Allele Frequency threshold for QC (default: 0.01)\nmissing_value: Value representing missing data (default: 9.0)\ncenter: Whether to center the data (default: true)\n\n\n\n\n\n","category":"type"},{"location":"manual/public/#NNMM.Equation","page":"Public Functions","title":"NNMM.Equation","text":"Equation\n\nA struct representing an equation connecting layers in NNMM.\n\nFields\n\nfrom_layer_name: Source layer name\nto_layer_name: Target layer name\nequation: Model equation string\nomics_name: Names of omics variables (or false if not applicable)\nphenotype_name: Names of phenotype variables (or false if not applicable)\ncovariate: Covariate variable names\nrandom: Random effect specifications\nactivation_function: Activation function (\"linear\", \"sigmoid\", \"tanh\", etc.)\npartial_connect_structure: Structure for partial connectivity\nstarting_value: Starting values for MCMC\nmethod: Bayesian method (\"BayesC\", \"BayesA\", \"BayesB\", etc.)\nPi: Prior probability of zero effect\nestimatePi: Whether to estimate Pi\nVariance parameters for G (genetic) and R (residual)\n\n\n\n\n\n","category":"type"},{"location":"manual/public/#NNMM.Omics","page":"Public Functions","title":"NNMM.Omics","text":"Omics\n\nA struct representing omics data (e.g., transcriptomics, metabolomics) in NNMM.\n\nFields\n\nname: Name identifier for this omics category\ntrait_names: Names for corresponding traits\nobsID: Row IDs for individuals\nfeatureID: Feature/omics variable IDs\nnObs: Number of observations\nnFeatures: Number of omics features\nnMarkers: Alias for nFeatures (compatibility)\ncentered: Whether data is centered\ndata: The omics data matrix\nntraits: Number of traits in the model\nAdditional fields for Bayesian MCMC computations\n\n\n\n\n\n","category":"type"},{"location":"manual/public/#NNMM.Phenotypes","page":"Public Functions","title":"NNMM.Phenotypes","text":"Phenotypes\n\nA struct representing phenotype data in NNMM.\n\nFields\n\nobsID: Individual IDs\nfeatureID: Phenotype/trait names\nnObs: Number of observations\nnPheno: Number of phenotypes\ndata: Phenotype data matrix\nnFeatures: Alias for nPheno\n\n\n\n\n\n","category":"type"},{"location":"manual/public/#NNMM.runNNMM","page":"Public Functions","title":"NNMM.runNNMM","text":"runNNMM(layers, equations; kwargs...)\n\nRun Neural Network Mixed Model (NNMM) analysis.\n\nArguments\n\nlayers: Vector of 3 Layer objects defining the network architecture\nLayer 1: Genotypes (SNPs)\nLayer 2: Omics/Latent traits\nLayer 3: Phenotypes\nequations: Vector of 2 Equation objects defining relationships\nEquation 1: Genotypes → Omics (1→2)\nEquation 2: Omics → Phenotypes (2→3)\n\nKeyword Arguments\n\nMCMC Settings\n\nchain_length::Integer=100: Total MCMC iterations\nburnin::Integer=0: Number of burn-in iterations to discard\noutput_samples_frequency::Integer: Save every nth sample (default: auto)\nupdate_priors_frequency::Integer=0: Update prior parameters every n iterations\n\nOutput Settings\n\noutputEBV=true: Output estimated breeding values\noutput_heritability=true: Calculate heritability estimates\noutput_folder=\"nnmm_results\": Directory for output files\n\nComputational Settings\n\nseed=false: Random seed for reproducibility\ndouble_precision=false: Use Float64 instead of Float32\nbig_memory=false: Enable memory-intensive optimizations\n\nReturns\n\nA dictionary containing:\n\nPosterior means for all parameters\nMCMC samples (saved to files)\nEBV estimates\n\nExample\n\n# Define layers\nlayer1 = Layer(name=\"geno\", file=\"genotypes.csv\")\nlayer2 = Layer(name=\"omics\", file=\"omics.csv\")\nlayer3 = Layer(name=\"pheno\", file=\"phenotypes.csv\")\n\n# Define equations\neq1 = Equation(\"omics = intercept + geno\", \n               method=\"BayesC\", omics_name=[\"o1\",\"o2\"])\neq2 = Equation(\"pheno = intercept + omics\", \n               activation_function=tanh, phenotype_name=[\"y\"])\n\n# Run analysis\nresults = runNNMM([layer1, layer2, layer3], [eq1, eq2],\n                  chain_length=5000, burnin=1000)\n\nSee also: Layer, Equation, describe\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#DataAPI.describe","page":"Public Functions","title":"DataAPI.describe","text":"describe(model::MME)\n\nPrint a summary of the mixed model equations (MME) structure.\n\nDisplays:\n\nModel equations (truncated if >5)\nTerm information (classification, fixed/random, number of levels)\nPrior distributions and hyperparameters\nVariance component settings\nMarker effect settings (if genomic data included)\n\nArguments\n\nmodel::MME: The mixed model equations object (created internally by runNNMM)\n\nOutput\n\nPrints formatted model summary to stdout including:\n\nModel equations\nTerm classification (covariate/factor, fixed/random)\nPrior settings for variance components\nMCMC configuration\n\nExample\n\n# After running runNNMM, the describe function is called automatically.\n# For manual inspection:\ndescribe(results[\"mme\"])\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.nnmm_get_genotypes","page":"Public Functions","title":"NNMM.nnmm_get_genotypes","text":"nnmm_get_genotypes(file, G=false; kwargs...)\n\nRead genotype data for the NNMM model (layer 1).\n\nArguments\n\nfile: Genotype data source - can be:\nString: Path to CSV file\nDataFrame: DataFrame with ID column + marker columns\nMatrix: Numeric matrix of genotypes\nG: Prior genetic variance (default: estimated from data)\n\nKeyword Arguments\n\nmethod: Bayesian method (\"BayesA\", \"BayesB\", \"BayesC\", \"RR-BLUP\", \"GBLUP\", \"BayesL\")\nPi: Prior inclusion probability (default: 0.0)\nestimatePi: Whether to estimate Pi (default: true)\nG_is_marker_variance: If true, G is marker variance; if false, G is genetic variance\ndf: Degrees of freedom for prior (default: 4.0)\nquality_control: Perform QC filtering (default: true)\nMAF: Minor allele frequency threshold (default: 0.01)\ncenter: Center genotypes (default: true)\nseparator: File delimiter (default: ',')\nheader: File has header row (default: true)\n\nReturns\n\nGenotypes struct containing processed genotype data and method parameters.\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.get_genotypes","page":"Public Functions","title":"NNMM.get_genotypes","text":"get_genotypes(args...; kwargs...)\n\nAlias for nnmm_get_genotypes. See that function for full documentation.\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.nnmm_get_omics","page":"Public Functions","title":"NNMM.nnmm_get_omics","text":"nnmm_get_omics(file, G=false; kwargs...)\n\nRead omics data for the NNMM model (layer 2 / middle layer).\n\nArguments\n\nfile: Path to CSV file containing omics data\nG: Prior genetic variance (default: estimated from data)\n\nKeyword Arguments\n\nomics_name: Vector of column names to use as omics features (required)\nmethod: Bayesian method (\"BayesA\", \"BayesB\", \"BayesC\", \"RR-BLUP\", \"BayesL\")\nPi: Prior inclusion probability (default: 0.0)\nestimatePi: Whether to estimate Pi (default: true)\nG_is_marker_variance: If true, G is marker variance; if false, G is genetic variance\ndf: Degrees of freedom for prior (default: 4.0)\nconstraint: Use independent variances for multi-trait (default: true)\nseparator: File delimiter (default: ',')\nheader: File has header row (default: true)\nmissing_value: String/value representing missing data (default: false)\n\nReturns\n\nOmics struct containing the omics data and method parameters.\n\nExample\n\nomics = nnmm_get_omics(\"omics_data.csv\", \n                       omics_name=[\"gene1\", \"gene2\", \"gene3\"],\n                       missing_value=\"NA\")\n\nNotes\n\nFirst column must be individual IDs\nMissing values will be sampled during MCMC (HMC for latent traits)\nUnlike genotypes, no quality control (MAF filtering) is applied\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.read_phenotypes","page":"Public Functions","title":"NNMM.read_phenotypes","text":"read_phenotypes(file; separator=',', header=true, missing_value=\"NA\", output_folder=\".\")\n\nRead phenotype data from a CSV file.\n\nArguments\n\nfile: Path to CSV file containing phenotype data\n\nKeyword Arguments\n\nseparator: Column delimiter (default: ',')\nheader: Whether file has header row (default: true)\nmissing_value: String representing missing values (default: \"NA\")\noutput_folder: Directory for output files (default: current directory)\n\nReturns\n\nDataFrame with phenotype data. First column is individual IDs.\n\nExample\n\npheno_df = read_phenotypes(\"phenotypes.csv\", missing_value=\"NA\")\n\nNotes\n\nFirst column must contain individual IDs\nMissing values are converted to Julia's missing type\nA file IDs_for_individuals_with_phenotypes.txt is written to output_folder\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.GWAS","page":"Public Functions","title":"NNMM.GWAS","text":"GWAS(marker_effects_file; header=true)\n\nCompute the model frequency for each marker.\n\nModel frequency is the probability that a marker is included in the model (i.e., has non-zero effect) across MCMC samples.\n\nArguments\n\nmarker_effects_file: Path to CSV file with MCMC samples of marker effects\nheader: Whether file has header row (default: true)\n\nReturns\n\nDataFrame with columns:\n\nmarker_ID: Marker identifier\nmodelfrequency: Proportion of samples where effect ≠ 0\n\nExample\n\n# After running NNMM with BayesB or BayesC\nfreq = GWAS(\"nnmm_results/MCMC_samples_marker_effects_genotypes.txt\")\n\n\n\n\n\nGWAS(model,map_file,marker_effects_file...;\n     window_size = \"1 Mb\",sliding_window = false,\n     GWAS = true, threshold = 0.001,\n     genetic_correlation = false,\n     header = true)\n\nrun genomic window-based GWAS\n\nMCMC samples of marker effects are stored in markereffectsfile with delimiter ','.\nmodel is either the model::MME used in analysis or the genotype cavariate matrix M::Array\nmap_file has the (sorted) marker position information with delimiter ','. If the map file is not provided, i.e., map_file=false, a fake map file will be generated with window_size markers in each 1 Mb window, and each 1 Mb window will be tested.\nIf two markereffectsfile are provided, and genetic_correlation = true, genomic correlation for each window is calculated.\nStatistics are computed for nonoverlapping windows of size window_size by default. If sliding_window = true, those for overlapping sliding windows are calculated.\nmap file format:\n\nmarkerID,chromosome,position\nm1,1,16977\nm2,1,434311\nm3,1,1025513\nm4,2,70350\nm5,2,101135\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.getEBV","page":"Public Functions","title":"NNMM.getEBV","text":"getEBV(model::MME,traiti)\n\n(internal function) Get breeding values for individuals defined by outputEBV(), defaulting to all genotyped individuals. This function is used inside MCMC functions for one MCMC samples from posterior distributions. e.g., non-NNBayespartial (multi-classs Bayes) : y1=M1α1[1]+M2α2[1]+M3α3[1]                                            y2=M1α1[2]+M2α2[2]+M3α3[2]; NNBayespartial:     y1=M1α1[1]                      y2=M2α2[1]                      y3=M3*α3[1];\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.Datasets.dataset","page":"Public Functions","title":"NNMM.Datasets.dataset","text":"dataset(file_name::AbstractString; dataset_name::AbstractString=\"\")\n\nGet the path to a built-in dataset file.\n\nArguments\n\nfile_name::AbstractString: The name of the file to retrieve\ndataset_name::AbstractString=\"\": Optional subdirectory name within the data folder\n\nReturns\n\nString: Full path to the requested data file\n\nExamples\n\nphenofile = dataset(\"phenotypes.csv\")\ngenofile = dataset(\"genotypes.txt\", dataset_name=\"example\")\n\n\n\n\n\n","category":"function"},{"location":"manual/public/#NNMM.PedModule.get_pedigree","page":"Public Functions","title":"NNMM.PedModule.get_pedigree","text":"get_pedigree(pedfile::AbstractString;header=false,separator=',',missingstrings=[\"0\"],output_folder=\".\")\n\nGet pedigree informtion from a pedigree file with header (defaulting to false) , separator (defaulting to ,) and missing values (defaulting to [\"0\"])\noutput_folder: Directory to save diagnostic files (defaulting to current directory)\nPedigree file format:\n\na,0,0\nc,a,b\nd,a,c\n\n\n\n\n\n","category":"function"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/#Partial-connected-neural-networks","page":"Part 4. Partial connected neural network","title":"Partial-connected neural networks","text":"In partial-connected neural networks, SNPs can be divided into groups by users, and each group connects to its own intermediate trait in the middle layer.  Genotype groups should be loaded seperatly.","category":"section"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/#example(c):-partial-connected-neural-networks,-all-intemediate-traits-are-unobserved","page":"Part 4. Partial connected neural network","title":"example(c): partial-connected neural networks, all intemediate traits are unobserved","text":"nonlinear function (to define relationship between middle layer and phenotye): tanh\nnumber of nodes in the middle layer: 3\nBayesian model (to sample marker effects on intemediate traits): \ngenotype group 1: single-trait BayesA\ngenotype group 2: single-trait BayesB\ngenotype group 3: single-trait BayesC\nsample the unobserved intemediate traits in the middle layer: Hamiltonian Monte Carlo\n\n(Image: )\n\n# Step 1: Load packages\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets,Random\nRandom.seed!(123)\n\n# Step 2: Read data\nphenofile   = dataset(\"phenotypes.csv\")       \ngenofile1   = dataset(\"genotypes_group1.csv\")  #path of genotype group1\ngenofile2   = dataset(\"genotypes_group2.csv\")  #path of genotype group2\ngenofile3   = dataset(\"genotypes_group3.csv\")  #path of genotype group3\n\nphenotypes = CSV.read(phenofile,DataFrame,delim = ',',header=true,missingstrings=[\"NA\"])\ngeno1  = get_genotypes(genofile1,separator=',',method=\"BayesA\");   #read genotype group1\ngeno2  = get_genotypes(genofile2,separator=',',method=\"BayesB\");   #read genotype group2\ngeno3  = get_genotypes(genofile3,separator=',',method=\"BayesC\");   #read genotype group3\n\n# Step 3: Build Model Equations\nmodel_equation = \"y1 = intercept + geno1 + geno2 + geno3\";  #middle node1=intercept + geno1\n                                                            #middle node2=intercept + geno2\n                                                            #middle node3=intercept + geno3\nmodel = build_model(model_equation,nonlinear_function=\"tanh\")\n\n# Step 4: Run Analysis\nout = runMCMC(model, phenotypes, chain_length=5000);\n\n# Step 5: Check Accuruacy\nresults    = innerjoin(out[\"EBV_NonLinear\"], phenotypes, on = :ID)\naccuruacy  = cor(results[!,:EBV],results[!,:bv1])","category":"section"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/#example(o3):-partial-connected-neural-networks-with-intermediate-omics-data","page":"Part 4. Partial connected neural network","title":"example(o3): partial-connected neural networks with intermediate omics data","text":"Same as for fully-connected neural network with intermediate omics data, the names of omics features should be put in the build_model() function through the latent_traits argument. The order of omics and the order of genotype groups in the model equation should be consistant.\n\nIn below example, we assume genotype group i only affect omics i (here i =1,...,5).\n\n(Image: )\n\n# Step 1: Load packages\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets,Random,HTTP \nRandom.seed!(1)\n\n# Step 2: Read data\nphenofile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/y.csv\").body\nomicsfile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/omics.csv\").body\n\ngenofile1   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group1.csv\").body\ngenofile2   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group2.csv\").body\ngenofile3   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group3.csv\").body\ngenofile4   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group4.csv\").body\ngenofile5   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group5.csv\").body\n\nphenotypes  = CSV.read(phenofile,DataFrame)\nomics       = CSV.read(omicsfile,DataFrame)[:,1:6] #only use first 5 omics for demonstration\nomics_names = names(omics)[2:end] #get names of omics\ninsertcols!(omics,2,:y => phenotypes[:,:y], :bv => phenotypes[:,:bv]) #phenotype and omics should be in the same dataframe\n\ngeno1_df    = CSV.read(genofile1,DataFrame)\ngeno2_df    = CSV.read(genofile2,DataFrame)\ngeno3_df    = CSV.read(genofile3,DataFrame)\ngeno4_df    = CSV.read(genofile4,DataFrame)\ngeno5_df    = CSV.read(genofile5,DataFrame)\n\ngeno1  = get_genotypes(geno1_df,separator=',',method=\"BayesA\");\ngeno2  = get_genotypes(geno2_df,separator=',',method=\"BayesB\");\ngeno3  = get_genotypes(geno3_df,separator=',',method=\"BayesC\");\ngeno4  = get_genotypes(geno4_df,separator=',',method=\"RR-BLUP\");\ngeno5  = get_genotypes(geno5_df,separator=',',method=\"BayesL\");\n\n\n# Step 3: Build Model Equations\nmodel_equation = \"y = intercept + geno1 + geno2 + geno3 + geno4 + geno5\"; #omics1=intercept+geno1; omics2=intercept+geno2; ...\nmodel = build_model(model_equation,\n\t\t    num_hidden_nodes=5,\n\t\t    nonlinear_function=\"sigmoid\",\n\t            latent_traits=omics_names)\n\n# Step 4: Run Analysis\nout = runMCMC(model, omics, chain_length=5000, printout_model_info=false);\n\n# Step 5: Check Accuruacy\nresults    = innerjoin(out[\"EBV_NonLinear\"], omics, on = :ID)\naccuruacy  = cor(results[!,:EBV],results[!,:bv])","category":"section"},{"location":"citing/citing/#Citing","page":"Citing","title":"Citing","text":"","category":"section"},{"location":"citing/citing/#Software","page":"Citing","title":"Software","text":"If you use NNMM.jl, please cite:\n\nZhao, T., Zeng, J., & Cheng, H. (2022). Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. GENETICS, iyac034. https://doi.org/10.1093/genetics/iyac034\nZhao, T., Fernando, R., & Cheng, H. (2021). Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. G3 Genes|Genomes|Genetics, jkab228. https://doi.org/10.1093/g3journal/jkab228","category":"section"},{"location":"citing/citing/#Methods","page":"Citing","title":"Methods","text":"","category":"section"},{"location":"citing/citing/#Neural-Network-Mixed-Models-(NNMM)","page":"Citing","title":"Neural Network Mixed Models (NNMM)","text":"Zhao T, Zeng J, Cheng H. 2022. Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. GENETICS, iyac034.\nZhao T, Fernando R, Cheng H. 2021. Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. G3 Genes|Genomes|Genetics, jkab228.","category":"section"},{"location":"citing/citing/#Bayesian-Alphabet-Methods","page":"Citing","title":"Bayesian Alphabet Methods","text":"Meuwissen T, Hayes B, Goddard M. 2001. Prediction of total genetic value using genome-wide dense marker maps. Genetics, 157:1819–1829.\nHabier D, Fernando R, Kizilkaya K, Garrick D. 2011. Extension of the bayesian alphabet for genomic selection. BMC Bioinformatics, 12(1), 186.\nCheng H, Garrick D, Fernando R. 2015. A fast and efficient Gibbs sampler for BayesB in whole-genome analyses. Genet Sel Evol, 47:80.\nCheng H, Kizilkaya K, Zeng J, Garrick D, Fernando R. 2018. Genomic Prediction from Multiple-Trait Bayesian Regression Methods Using Mixture Priors. Genetics, 209(1): 89-103.","category":"section"},{"location":"citing/citing/#Single-step-Methods","page":"Citing","title":"Single-step Methods","text":"Fernando R, Dekkers J, Garrick D. 2015. A class of Bayesian methods to combine large numbers of genotyped and non-genotyped animals for whole-genome analyses. Genetics Selection Evolution, 46(1), 50.\nFernando R, Cheng H, Golden B, Garrick D. 2016. Computational strategies for alternative single-step Bayesian regression models with large numbers of genotyped and non-genotyped animals. Genetics Selection Evolution, 48(1), 96.","category":"section"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/#User-defined-relationship-between-middle-layer-(intermediate-traits)-and-output-layer-(phenotype)","page":"Part 5. User-defined nonlinear function","title":"User-defined relationship between middle layer (intermediate traits) and output layer (phenotype)","text":"Firstly, a function should be pre-defined by user, where each function argument represents the input from a intermediate trait. For example, with two omics, we can have pig_growth(omics1,omics2) = sqrt(omics1^2 / (omics1^2 + omics2^2)))\nThen, put the name of the user-define function in the nonlinear_function arugument when build the model.","category":"section"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/#example(b):-user-defined-relationship-between-middle-layer-(intermediate-traits)-and-output-layer-(phenotype)","page":"Part 5. User-defined nonlinear function","title":"example(b): user-defined relationship between middle layer (intermediate traits) and output layer (phenotype)","text":"number of nodes in the middle layer: 2\nnonlinear function (to define relationship between middle nodes and phenotype): y = sqrt(x1^2 / (x1^2 + x2^2))\nsample the missing omics in the middle layer: Matropolis-Hastings\n\nAll the other code are the same as before, except Step 3: Build Model Equations. Note that user-defined relationship is supported in both fully-connected and partial-connected neural network. \n\n# Step 3: Build Model Equations\npig_growth(x1,x2) = sqrt(x1^2 / (x1^2 + x2^2))\nmodel = build_model(model_equation,\n                    num_hidden_nodes=2,\n                    nonlinear_function=pig_growth);","category":"section"},{"location":"manual/tutorial/#Step-by-Step-Tutorial","page":"Tutorial","title":"Step-by-Step Tutorial","text":"This tutorial walks through a complete NNMM analysis using simulated data.","category":"section"},{"location":"manual/tutorial/#Overview","page":"Tutorial","title":"Overview","text":"We'll perform genomic prediction using a two-layer Neural Network Mixed Model:\n\nGenotypes (SNPs) → Omics (gene expression) → Phenotype (trait)\n\nThis architecture allows us to:\n\nModel the biological pathway from DNA to gene expression to phenotype\nHandle missing omics data via HMC sampling\nObtain estimated breeding values (EBV) for selection","category":"section"},{"location":"manual/tutorial/#Step-1:-Load-Packages-and-Data","page":"Tutorial","title":"Step 1: Load Packages and Data","text":"using NNMM\nusing NNMM.Datasets\nusing DataFrames\nusing CSV\n\n# Load built-in simulated dataset\ngeno_path = Datasets.dataset(\"genotypes_1000snps.txt\", dataset_name=\"simulated_omics_data\")\npheno_path = Datasets.dataset(\"phenotypes_sim.txt\", dataset_name=\"simulated_omics_data\")\nped_path = Datasets.dataset(\"pedigree.txt\", dataset_name=\"simulated_omics_data\")\n\n# Check the data\nprintln(\"Genotype file: \", geno_path)\nprintln(\"Phenotype file: \", pheno_path)","category":"section"},{"location":"manual/tutorial/#Step-2:-Prepare-Data-Files","page":"Tutorial","title":"Step 2: Prepare Data Files","text":"The phenotype file contains both omics and trait data. We need to split them:\n\n# Read the combined phenotype file\npheno_df = CSV.read(pheno_path, DataFrame)\nprintln(\"Columns: \", names(pheno_df))\n\n# Create omics file (use 5 omics features)\nomics_df = pheno_df[:, [:ID, :omic1, :omic2, :omic3, :omic4, :omic5]]\nomics_file = \"omics_data.csv\"\nCSV.write(omics_file, omics_df; missingstring=\"NA\")\n\n# Create phenotype file (single trait)\ntrait_df = pheno_df[:, [:ID, :trait1]]\ntrait_file = \"phenotypes.csv\"\nCSV.write(trait_file, trait_df; missingstring=\"NA\")\n\nprintln(\"Created: $omics_file and $trait_file\")","category":"section"},{"location":"manual/tutorial/#Step-3:-Define-the-Network-Architecture","page":"Tutorial","title":"Step 3: Define the Network Architecture","text":"NNMM uses Layer and Equation structs to define the model:\n\n# Define the three layers\nlayers = [\n    # Layer 1: Genotypes (input layer)\n    Layer(\n        layer_name = \"genotypes\",\n        data_path = [geno_path]\n    ),\n    \n    # Layer 2: Omics (hidden/intermediate layer)\n    Layer(\n        layer_name = \"omics\",\n        data_path = omics_file,\n        missing_value = \"NA\"    # Handle missing omics values\n    ),\n    \n    # Layer 3: Phenotypes (output layer)\n    Layer(\n        layer_name = \"phenotypes\",\n        data_path = trait_file,\n        missing_value = \"NA\"\n    )\n]","category":"section"},{"location":"manual/tutorial/#Step-4:-Define-the-Equations","page":"Tutorial","title":"Step 4: Define the Equations","text":"Equations connect the layers and specify the statistical method:\n\nequations = [\n    # Layer 1 → Layer 2: Genotypes predict omics\n    Equation(\n        from_layer_name = \"genotypes\",\n        to_layer_name = \"omics\",\n        equation = \"omics = intercept + genotypes\",\n        omics_name = [\"omic1\", \"omic2\", \"omic3\", \"omic4\", \"omic5\"],\n        method = \"BayesC\",        # Bayesian method\n        estimatePi = true         # Estimate marker inclusion probability\n    ),\n    \n    # Layer 2 → Layer 3: Omics predict phenotype\n    Equation(\n        from_layer_name = \"omics\",\n        to_layer_name = \"phenotypes\",\n        equation = \"phenotypes = intercept + omics\",\n        phenotype_name = [\"trait1\"],\n        method = \"BayesC\",\n        estimatePi = true\n    )\n]","category":"section"},{"location":"manual/tutorial/#Step-5:-Run-the-Analysis","page":"Tutorial","title":"Step 5: Run the Analysis","text":"# Run NNMM with MCMC sampling\nresults = runNNMM(\n    layers, \n    equations;\n    chain_length = 5000,         # Number of MCMC iterations\n    burnin = 1000,               # Burn-in iterations\n    printout_frequency = 1000,   # Print progress every N iterations\n    output_folder = \"nnmm_output\"  # Where to save results\n)\n\ntip: For quick testing\nUse chain_length=100 for testing. For production, use at least chain_length=50000.","category":"section"},{"location":"manual/tutorial/#Step-6:-Examine-Results","page":"Tutorial","title":"Step 6: Examine Results","text":"","category":"section"},{"location":"manual/tutorial/#Estimated-Breeding-Values-(EBV)","page":"Tutorial","title":"Estimated Breeding Values (EBV)","text":"# Get EBV for all individuals\nebv = results[\"EBV_NonLinear\"]\nprintln(\"EBV for first 10 individuals:\")\ndisplay(first(ebv, 10))\n\n# Summary statistics\nprintln(\"\\nEBV summary:\")\nprintln(\"Mean: \", round(mean(ebv.EBV), digits=3))\nprintln(\"Std:  \", round(std(ebv.EBV), digits=3))\nprintln(\"Min:  \", round(minimum(ebv.EBV), digits=3))\nprintln(\"Max:  \", round(maximum(ebv.EBV), digits=3))","category":"section"},{"location":"manual/tutorial/#MCMC-Output-Files","page":"Tutorial","title":"MCMC Output Files","text":"The output folder contains MCMC samples for:\n\nFile Description\nMCMC_samples_marker_effects_*.txt Marker effect samples\nMCMC_samples_marker_effects_variances_*.txt Marker variance samples\nMCMC_samples_residual_variance.txt Residual variance samples\nMCMC_samples_pi_*.txt Marker inclusion probability samples\nMCMC_samples_EBV_*.txt EBV samples per trait/omics\nMCMC_samples_EPV_NonLinear.txt Final phenotypic value samples","category":"section"},{"location":"manual/tutorial/#Step-7:-Post-Analysis-with-GWAS","page":"Tutorial","title":"Step 7: Post-Analysis with GWAS","text":"Identify important markers using model frequency:\n\n# GWAS based on model frequency (probability marker is included)\nmarker_file = \"nnmm_output/MCMC_samples_marker_effects_genotypes_omic1.txt\"\ngwas_result = GWAS(marker_file)\n\n# Top markers for omic1\nsorted_gwas = sort(gwas_result, :modelfrequency, rev=true)\nprintln(\"Top 10 markers for omic1:\")\ndisplay(first(sorted_gwas, 10))","category":"section"},{"location":"manual/tutorial/#Complete-Script","page":"Tutorial","title":"Complete Script","text":"Here's the complete analysis in one script:\n\nusing NNMM\nusing NNMM.Datasets\nusing DataFrames\nusing CSV\n\n# === Data Preparation ===\ngeno_path = Datasets.dataset(\"genotypes_1000snps.txt\", dataset_name=\"simulated_omics_data\")\npheno_path = Datasets.dataset(\"phenotypes_sim.txt\", dataset_name=\"simulated_omics_data\")\n\npheno_df = CSV.read(pheno_path, DataFrame)\n\n# Create omics and trait files\nomics_df = pheno_df[:, [:ID, :omic1, :omic2, :omic3]]\nCSV.write(\"omics.csv\", omics_df; missingstring=\"NA\")\n\ntrait_df = pheno_df[:, [:ID, :trait1]]\nCSV.write(\"traits.csv\", trait_df; missingstring=\"NA\")\n\n# === Define Model ===\nlayers = [\n    Layer(layer_name=\"geno\", data_path=[geno_path]),\n    Layer(layer_name=\"omics\", data_path=\"omics.csv\", missing_value=\"NA\"),\n    Layer(layer_name=\"pheno\", data_path=\"traits.csv\", missing_value=\"NA\")\n]\n\nequations = [\n    Equation(\n        from_layer_name=\"geno\", to_layer_name=\"omics\",\n        equation=\"omics = intercept + geno\",\n        omics_name=[\"omic1\", \"omic2\", \"omic3\"],\n        method=\"BayesC\"\n    ),\n    Equation(\n        from_layer_name=\"omics\", to_layer_name=\"pheno\",\n        equation=\"pheno = intercept + omics\",\n        phenotype_name=[\"trait1\"],\n        method=\"BayesC\"\n    )\n]\n\n# === Run MCMC ===\nresults = runNNMM(layers, equations; \n                  chain_length=5000, \n                  burnin=1000,\n                  output_folder=\"results\")\n\n# === Results ===\nebv = results[\"EBV_NonLinear\"]\nprintln(\"Analysis complete! EBV saved in results folder.\")\nprintln(\"Mean EBV: \", mean(ebv.EBV))\n\n# Cleanup temp files\nrm(\"omics.csv\", force=true)\nrm(\"traits.csv\", force=true)","category":"section"},{"location":"manual/tutorial/#Next-Steps","page":"Tutorial","title":"Next Steps","text":"Different methods: Try method=\"BayesA\", \"BayesB\", \"BayesL\", or \"RR-BLUP\"\nAdd pedigree: Include pedigree_file in Layer for relationship matrices\nMulti-trait: Add more phenotypes in phenotype_name\nActivation functions: Use activation_function=\"tanh\" for non-linear relationships\nMissing omics: NNMM automatically handles missing omics via HMC sampling\n\nSee the NNMM Examples for more advanced use cases.","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Part-6.-NNMM-as-Traditional-Genomic-Prediction","page":"Part 6. Traditional genomic prediction","title":"Part 6. NNMM as Traditional Genomic Prediction","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Overview","page":"Part 6. Traditional genomic prediction","title":"Overview","text":"NNMM is a general framework that encompasses traditional genomic prediction models as special cases. By configuring the network architecture appropriately, you can use NNMM to perform standard single-trait or multi-trait BayesC (and other Bayesian alphabet methods).\n\nThis demonstrates that NNMM generalizes traditional genomic prediction rather than replacing it.","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#The-Key-Insight","page":"Part 6. Traditional genomic prediction","title":"The Key Insight","text":"Traditional genomic prediction models like BayesC can be expressed as a special case of NNMM by:\n\nUsing a single completely missing node in the middle layer - This latent node will be sampled based on genotype information\nUsing a linear activation function - This ensures the relationship between the latent node and phenotype is linear\n\n┌─────────────┐     ┌─────────────────┐     ┌─────────────┐\n│  Genotypes  │ ──► │  Latent Nodes   │ ──► │ Phenotype   │\n│   (SNPs)    │     │ (100% missing)  │     │    (y)      │\n└─────────────┘     └─────────────────┘     └─────────────┘\n                    Sampled via BayesC       Linear activation\n\nThis is mathematically equivalent to the traditional model:\n\ny = Xbeta + Zalpha + e\n\nwhere alpha are marker effects sampled using BayesC.","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Example:-Single-Trait-BayesC","page":"Part 6. Traditional genomic prediction","title":"Example: Single-Trait BayesC","text":"using NNMM, DataFrames, CSV\n\n# Step 1: Create a middle layer with completely missing latent nodes\n# Note: Use at least 2 latent nodes for matrix operations in HMC\nomics_df = DataFrame(\n    ID = individual_ids,\n    latent1 = fill(missing, n_individuals),  # All NA\n    latent2 = fill(missing, n_individuals)   # All NA\n)\nCSV.write(\"latent.csv\", omics_df; missingstring=\"NA\")\n\n# Step 2: Define 3-layer network\nlayers = [\n    Layer(layer_name=\"geno\", data_path=[\"genotypes.csv\"]),\n    Layer(layer_name=\"latent\", data_path=\"latent.csv\", missing_value=\"NA\"),\n    Layer(layer_name=\"phenotypes\", data_path=\"phenotypes.csv\", missing_value=\"NA\")\n]\n\n# Step 3: Define equations with LINEAR activation\nequations = [\n    # Genotypes → Latent (marker regression)\n    Equation(\n        from_layer_name=\"geno\",\n        to_layer_name=\"latent\",\n        equation=\"latent = intercept + geno\",\n        omics_name=[\"latent1\", \"latent2\"],\n        method=\"BayesC\",\n        estimatePi=true\n    ),\n    # Latent → Phenotype (LINEAR activation = standard regression)\n    Equation(\n        from_layer_name=\"latent\",\n        to_layer_name=\"phenotypes\",\n        equation=\"phenotypes = intercept + latent\",\n        phenotype_name=[\"y1\"],\n        method=\"BayesC\",\n        activation_function=\"linear\"  # Key: linear activation\n    )\n]\n\n# Step 4: Run analysis\nresult = runNNMM(layers, equations; chain_length=50000, burnin=10000)","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Note-on-Multi-Trait-Analysis","page":"Part 6. Traditional genomic prediction","title":"Note on Multi-Trait Analysis","text":"note: Current Limitation\nMultiple phenotypes in the output layer are not yet supported in the current NNMM implementation. Multi-trait genomic prediction will be supported in a future release.","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Comparison-with-Direct-BayesC","page":"Part 6. Traditional genomic prediction","title":"Comparison with Direct BayesC","text":"Aspect Traditional BayesC NNMM as BayesC\nModel equation y = intercept + geno 3-layer network with missing middle\nMarker effects Directly estimated Estimated via latent layer\nFlexibility Fixed model Can add intermediate omics later\nResults EBV, marker effects EBV, marker effects","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Why-Use-NNMM-for-Traditional-Analysis?","page":"Part 6. Traditional genomic prediction","title":"Why Use NNMM for Traditional Analysis?","text":"Unified framework: Same code structure for simple and complex models\nEasy extension: Add observed omics data to the middle layer when available\nGradual complexity: Start with traditional model, add neural network components as needed\nFuture-proof: As more omics data becomes available, simply fill in the missing values","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Other-Bayesian-Methods","page":"Part 6. Traditional genomic prediction","title":"Other Bayesian Methods","text":"You can use any Bayesian alphabet method by changing the method parameter:\n\n# BayesA\nEquation(..., method=\"BayesA\")\n\n# BayesB  \nEquation(..., method=\"BayesB\", estimatePi=true)\n\n# BayesL (Bayesian LASSO)\nEquation(..., method=\"BayesL\")\n\n# RR-BLUP (Ridge Regression)\nEquation(..., method=\"RR-BLUP\")","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Summary","page":"Part 6. Traditional genomic prediction","title":"Summary","text":"NNMM provides a superset of traditional genomic prediction methods:\n\nTraditional BayesC = NNMM with completely missing middle layer + linear activation\nNNMM with omics = Same framework with observed intermediate traits\nFull neural network = Same framework with nonlinear activation functions\n\nThis design allows researchers to start with familiar methods and gradually incorporate more complex models as data and understanding grow.","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Mixed-effect-neural-network:-Genotypes-(complete/incomplete)-Intemediate-omics-features-Phenotyes","page":"Part 3. NNMM with intermediate omics","title":"Mixed effect neural network: Genotypes -> (complete/incomplete) Intemediate omics features -> Phenotyes","text":"Tips:\n\nPut the names of omics features in the build_model() function through the latent_traits argument.\nIf there are many omics features (e.g., 1000), it is recommanded to set printout_model_info=false in the runMCMC() function.\nMissing omics data for individuals in the training dataset (i.e., individuals with phenotypes) is allowed. When you read a file with missing values via the CSV.read() function, the missingstrings argument should be used to set sentinel values that will be parsed as missing.\nFor individuals in the testing dataset (i.e., individuals without phenotypes), if the testing individuals have complete omics data, then incorporating the omics data of those individuals may improve the relationship between input layer (genotype) and middle layer (omics).","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#example(o1):-fully-connected-neural-networks-with-observed-intemediate-omics-features","page":"Part 3. NNMM with intermediate omics","title":"example(o1): fully-connected neural networks with observed intemediate omics features","text":"nonlinear function (to define relationship between omics and phenotye): sigmoid (other supported activation functions: \"tanh\", \"relu\", \"leakyrelu\", \"linear\")\nnumber of omics features in the middle layer: 10\nBayesian model: multiple independent single-trait BayesC (to sample marker effects on intemediate omics)\nsample the missing omics in the middle layer: Hamiltonian Monte Carlo\n\n(Image: )\n\n# Step 1: Load packages\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets, Random, HTTP #HTTP to download demo data from github\nRandom.seed!(123)\n\n# Step 2: Read data (from github)\nphenofile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/y.csv\").body\nomicsfile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/omics.csv\").body\ngenofile   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_n100_p200.csv\").body\nphenotypes = CSV.read(phenofile,DataFrame)\nomics      = CSV.read(omicsfile,DataFrame)\ngeno_df    = CSV.read(genofile,DataFrame)\n\nomics_names = names(omics)[2:end]  #get names of omics\ninsertcols!(omics,2,:y => phenotypes[:,:y], :bv => phenotypes[:,:bv]) #phenotype and omics should be in the same dataframe\n\ngenotypes = get_genotypes(geno_df,separator=',',method=\"BayesC\")\n\n# Step 3: Build Model Equations\nmodel_equation  =\"y = intercept + genotypes\" #name of phenotypes is \"y\"\n                                             #name of genotypes is \"genotypes\" (user-defined in the previous step)\n                                             #the single-trait mixed model used between input and each omics is: omics = intercept + genotypes\nmodel = build_model(model_equation,\n\t\t    num_hidden_nodes=10,          #number of omics in middle layer is 3\n                    latent_traits=omics_names,    #name of all omics\n\t\t    nonlinear_function=\"sigmoid\") #sigmoid function is used to approximate relationship between omics and phenotypes\n\n# Step 4: Run Analysis\nout=runMCMC(model, omics, chain_length=5000, printout_model_info=false);\n\n# Step 5: Check Accuruacy\nresults    = innerjoin(out[\"EBV_NonLinear\"], omics, on = :ID)\naccuruacy  = cor(results[!,:EBV],results[!,:bv])","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Includes-a-residual-that-is-not-mediated-by-other-omics-features","page":"Part 3. NNMM with intermediate omics","title":"Includes a residual that is not mediated by other omics features","text":"To include residuals polygenic component (i.e. directly from genotypes to phenotypes, not mediated by omics features), you can additional hidden nodes in the middle layer (see example (o2)). This can also be achieved in a partial-connected neural network in a same manner.\n\n(Image: )","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#example(o2):-fully-connected-neural-network-with-residuals","page":"Part 3. NNMM with intermediate omics","title":"example(o2): fully-connected neural network with residuals","text":"For all individuals, this extra hidden node will be treated as unknown to be sampled.\n\n# Step 1: Load packages\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets, Random, HTTP \nRandom.seed!(123)\n\n# Step 2: Read data (from github)\nphenofile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/y.csv\").body\nomicsfile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/omics.csv\").body\ngenofile   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_n100_p200.csv\").body\nphenotypes = CSV.read(phenofile,DataFrame)\nomics      = CSV.read(omicsfile,DataFrame)\ngeno_df    = CSV.read(genofile,DataFrame)\n\ninsertcols!(omics, :residual => missing)  #create a hidden node to account for residuals\nomics[!,:residual] = convert(Vector{Union{Missing,Float64}}, omics[!,:residual]) #transform the datatype is required for Julia\nomics_names = names(omics)[2:end]  #get names of 10 omics and 1 hidden node\ninsertcols!(omics,2,:y => phenotypes[:,:y], :bv => phenotypes[:,:bv]) #phenotype and omics should be in the same dataframe\n\ngenotypes = get_genotypes(geno_df,separator=',',method=\"BayesC\")\n\n# Step 3: Build Model Equations\nmodel_equation  =\"y = intercept + genotypes\" \nmodel = build_model(model_equation,\n\t\t    num_hidden_nodes=11,   #10 omcis and 1 hidden node\n                    latent_traits=omics_names,\n\t\t    nonlinear_function=\"sigmoid\")\n\n# Step 4: Run Analysis\nout = runMCMC(model,omics,chain_length=5000,printout_model_info=false)\n\n# Step 5: Check Accuruacy\nresults    = innerjoin(out[\"EBV_NonLinear\"], omics, on = :ID)\naccuruacy  = cor(results[!,:EBV],results[!,:bv])\n\nUsers can also add extra hidden nodes in the partial-connected neural network. Please check next documentation for building a partial-connected neural network.","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Output-files","page":"Part 3. NNMM with intermediate omics","title":"Output files","text":"Same as those described in Part2.","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Julia-Tips:","page":"Part 3. NNMM with intermediate omics","title":"Julia Tips:","text":"You may want to set missing values manually, for example, setting the phenotypes for individuals in testing dataset as missing. Firstly, the type of  columns should be changed to allow missing, e.g., phenotypes[!,:y] =  convert(Vector{Union{Missing,Float64}}, phenotypes[!,:y]). Then, missing can be set manually, e.g., phenotypes[10:11,:y1] .= missing forces the 10th and 11th elements to be missing.","category":"section"},{"location":"","page":"Home","title":"Home","text":"(Image: NNMM)","category":"section"},{"location":"#NNMM.jl-Neural-Network-Mixed-Models","page":"Home","title":"NNMM.jl - Neural Network Mixed Models","text":"NNMM.jl is a Julia package for Mixed Effect Neural Networks that extend traditional linear mixed models to multilayer neural networks for genomic prediction and genome-wide association studies.","category":"section"},{"location":"#Key-Features","page":"Home","title":"Key Features","text":"Generalizes Traditional Methods: Traditional BayesC and other Bayesian alphabet methods are special cases of NNMM\nNeural Network Architecture: Extend linear mixed models to multilayer neural networks\nIntermediate Omics Integration: Incorporate known intermediate omics features (e.g., gene expression) in the middle layer\nFlexible Missing Data Handling: Allow any patterns of missing data in intermediate layers\nBayesian Framework: Full Bayesian inference using MCMC and Hamiltonian Monte Carlo\nMulti-threaded Parallelism: Efficient parallel computing for large-scale analyses","category":"section"},{"location":"#Network-Architecture","page":"Home","title":"Network Architecture","text":"NNMM models relationships between:\n\nInput Layer: Genotypes (SNP markers)\nMiddle Layer: Intermediate traits (e.g., gene expression, metabolomics)\nOutput Layer: Phenotypes\n\nThe package supports:\n\nFully-connected neural networks\nPartial-connected neural networks\nUser-defined nonlinear activation functions","category":"section"},{"location":"#Installation","page":"Home","title":"Installation","text":"using Pkg\nPkg.add(\"NNMM\")\n\nOr for the latest development version:\n\nPkg.add(url=\"https://github.com/reworkhow/NNMM.jl\")","category":"section"},{"location":"#Quick-Start","page":"Home","title":"Quick Start","text":"using NNMM, CSV, DataFrames\n\n# Load data\nphenotypes = CSV.read(\"phenotypes.csv\", DataFrame)\nomics = CSV.read(\"omics.csv\", DataFrame)\ngenotypes = CSV.read(\"genotypes.csv\", DataFrame)\n\n# Define network layers\nlayer1 = Layer(data=genotypes)\nlayer2 = Layer(data=omics, n_latent=size(omics, 2)-1)\nlayer3 = Layer(data=phenotypes)\n\n# Build model equation\neq = Equation(layer1, layer2, layer3, \"y\")\n\n# Run MCMC\nresults = runNNMM(eq, chain_length=5000, output_folder=\"results\")","category":"section"},{"location":"#Supporting-and-Citing","page":"Home","title":"Supporting and Citing","text":"If you use NNMM.jl for your research, please cite:\n\nZhao, T., Zeng, J., & Cheng, H. (2022). Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. GENETICS. https://doi.org/10.1093/genetics/iyac034\n\nZhao, T., Fernando, R., & Cheng, H. (2021). Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. G3 Genes|Genomes|Genetics. https://doi.org/10.1093/g3journal/jkab228\n\nPlease star the repository here to help demonstrate community involvement.","category":"section"},{"location":"#Getting-Help","page":"Home","title":"Getting Help","text":"Open an issue on GitHub\nContact: <qtlcheng@ucdavis.edu>","category":"section"},{"location":"#Tutorials","page":"Home","title":"Tutorials","text":"","category":"section"},{"location":"#NNMM-Tutorials","page":"Home","title":"NNMM Tutorials","text":"Pages = [\n  \"nnmm/Part1_introduction.md\",\n  \"nnmm/Part2_NNMM.md\",\n  \"nnmm/Part3_NNMMwithIntermediateOmics.md\",\n  \"nnmm/Part4_PartialConnectedNeuralNetwork.md\",\n  \"nnmm/Part5_UserDefinedNonlinearFunction.md\",\n  \"nnmm/Part6_TraditionalGenomicPrediction.md\",\n]\nDepth = 2","category":"section"},{"location":"#Manual","page":"Home","title":"Manual","text":"Pages = [\n  \"manual/getstarted.md\",\n  \"manual/public.md\",\n]\nDepth = 2","category":"section"}]
}
