var documenterSearchIndex = {"docs":
[{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/#Partial-connected-neural-networks","page":"Part 4. Partial connected neural network","title":"Partial-connected neural networks","text":"","category":"section"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"In partial-connected neural networks, SNPs can be divided into groups by users, and each group connects to its own intermediate trait in the middle layer.  Genotype groups should be loaded seperatly.","category":"page"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/#example(c):-partial-connected-neural-networks,-all-intemediate-traits-are-unobserved","page":"Part 4. Partial connected neural network","title":"example(c): partial-connected neural networks, all intemediate traits are unobserved","text":"","category":"section"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"nonlinear function (to define relationship between middle layer and phenotye): tanh\nnumber of nodes in the middle layer: 3\nBayesian model (to sample marker effects on intemediate traits): \ngenotype group 1: single-trait BayesA\ngenotype group 2: single-trait BayesB\ngenotype group 3: single-trait BayesC\nsample the unobserved intemediate traits in the middle layer: Hamiltonian Monte Carlo","category":"page"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"(Image: )","category":"page"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"# Step 1: Load packages\r\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets,Random\r\nRandom.seed!(123)\r\n\r\n# Step 2: Read data\r\nphenofile   = dataset(\"phenotypes.csv\")       \r\ngenofile1   = dataset(\"genotypes_group1.csv\")  #path of genotype group1\r\ngenofile2   = dataset(\"genotypes_group2.csv\")  #path of genotype group2\r\ngenofile3   = dataset(\"genotypes_group3.csv\")  #path of genotype group3\r\n\r\nphenotypes = CSV.read(phenofile,DataFrame,delim = ',',header=true,missingstrings=[\"NA\"])\r\ngeno1  = get_genotypes(genofile1,separator=',',method=\"BayesA\");   #read genotype group1\r\ngeno2  = get_genotypes(genofile2,separator=',',method=\"BayesB\");   #read genotype group2\r\ngeno3  = get_genotypes(genofile3,separator=',',method=\"BayesC\");   #read genotype group3\r\n\r\n# Step 3: Build Model Equations\r\nmodel_equation = \"y1 = intercept + geno1 + geno2 + geno3\";  #middle node1=intercept + geno1\r\n                                                            #middle node2=intercept + geno2\r\n                                                            #middle node3=intercept + geno3\r\nmodel = build_model(model_equation,nonlinear_function=\"tanh\")\r\n\r\n# Step 4: Run Analysis\r\nout = runMCMC(model, phenotypes, chain_length=5000);\r\n\r\n# Step 5: Check Accuruacy\r\nresults    = innerjoin(out[\"EBV_NonLinear\"], phenotypes, on = :ID)\r\naccuruacy  = cor(results[!,:EBV],results[!,:bv1])","category":"page"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/#example(o3):-partial-connected-neural-networks-with-intermediate-omics-data","page":"Part 4. Partial connected neural network","title":"example(o3): partial-connected neural networks with intermediate omics data","text":"","category":"section"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"Same as for fully-connected neural network with intermediate omics data, the names of omics features should be put in the build_model() function through the latent_traits argument. The order of omics and the order of genotype groups in the model equation should be consistant.","category":"page"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"In below example, we assume genotype group i only affect omics i (here i =1,...,5).","category":"page"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"(Image: )","category":"page"},{"location":"nnmm/Part4_PartialConnectedNeuralNetwork/","page":"Part 4. Partial connected neural network","title":"Part 4. Partial connected neural network","text":"# Step 1: Load packages\r\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets,Random,HTTP \r\nRandom.seed!(1)\r\n\r\n# Step 2: Read data\r\nphenofile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/y.csv\").body\r\nomicsfile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/omics.csv\").body\r\n\r\ngenofile1   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group1.csv\").body\r\ngenofile2   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group2.csv\").body\r\ngenofile3   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group3.csv\").body\r\ngenofile4   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group4.csv\").body\r\ngenofile5   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_group5.csv\").body\r\n\r\nphenotypes  = CSV.read(phenofile,DataFrame)\r\nomics       = CSV.read(omicsfile,DataFrame)[:,1:6] #only use first 5 omics for demonstration\r\nomics_names = names(omics)[2:end] #get names of omics\r\ninsertcols!(omics,2,:y => phenotypes[:,:y], :bv => phenotypes[:,:bv]) #phenotype and omics should be in the same dataframe\r\n\r\ngeno1_df    = CSV.read(genofile1,DataFrame)\r\ngeno2_df    = CSV.read(genofile2,DataFrame)\r\ngeno3_df    = CSV.read(genofile3,DataFrame)\r\ngeno4_df    = CSV.read(genofile4,DataFrame)\r\ngeno5_df    = CSV.read(genofile5,DataFrame)\r\n\r\ngeno1  = get_genotypes(geno1_df,separator=',',method=\"BayesA\");\r\ngeno2  = get_genotypes(geno2_df,separator=',',method=\"BayesB\");\r\ngeno3  = get_genotypes(geno3_df,separator=',',method=\"BayesC\");\r\ngeno4  = get_genotypes(geno4_df,separator=',',method=\"RR-BLUP\");\r\ngeno5  = get_genotypes(geno5_df,separator=',',method=\"BayesL\");\r\n\r\n\r\n# Step 3: Build Model Equations\r\nmodel_equation = \"y = intercept + geno1 + geno2 + geno3 + geno4 + geno5\"; #omics1=intercept+geno1; omics2=intercept+geno2; ...\r\nmodel = build_model(model_equation,\r\n\t\t    num_hidden_nodes=5,\r\n\t\t    nonlinear_function=\"sigmoid\",\r\n\t            latent_traits=omics_names)\r\n\r\n# Step 4: Run Analysis\r\nout = runMCMC(model, omics, chain_length=5000, printout_model_info=false);\r\n\r\n# Step 5: Check Accuruacy\r\nresults    = innerjoin(out[\"EBV_NonLinear\"], omics, on = :ID)\r\naccuruacy  = cor(results[!,:EBV],results[!,:bv])","category":"page"},{"location":"citing/citing/#Citing","page":"Citing","title":"Citing","text":"","category":"section"},{"location":"citing/citing/#Software","page":"Citing","title":"Software","text":"","category":"section"},{"location":"citing/citing/","page":"Citing","title":"Citing","text":"If you use NNMM.jl, please cite:","category":"page"},{"location":"citing/citing/","page":"Citing","title":"Citing","text":"Zhao, T., Zeng, J., & Cheng, H. (2022). Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. GENETICS, iyac034. https://doi.org/10.1093/genetics/iyac034\nZhao, T., Fernando, R., & Cheng, H. (2021). Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. G3 Genes|Genomes|Genetics, jkab228. https://doi.org/10.1093/g3journal/jkab228","category":"page"},{"location":"citing/citing/#Methods","page":"Citing","title":"Methods","text":"","category":"section"},{"location":"citing/citing/#Neural-Network-Mixed-Models-(NNMM)","page":"Citing","title":"Neural Network Mixed Models (NNMM)","text":"","category":"section"},{"location":"citing/citing/","page":"Citing","title":"Citing","text":"Zhao T, Zeng J, Cheng H. 2022. Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. GENETICS, iyac034.\nZhao T, Fernando R, Cheng H. 2021. Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. G3 Genes|Genomes|Genetics, jkab228.","category":"page"},{"location":"citing/citing/#Bayesian-Alphabet-Methods","page":"Citing","title":"Bayesian Alphabet Methods","text":"","category":"section"},{"location":"citing/citing/","page":"Citing","title":"Citing","text":"Meuwissen T, Hayes B, Goddard M. 2001. Prediction of total genetic value using genome-wide dense marker maps. Genetics, 157:1819–1829.\nHabier D, Fernando R, Kizilkaya K, Garrick D. 2011. Extension of the bayesian alphabet for genomic selection. BMC Bioinformatics, 12(1), 186.\nCheng H, Garrick D, Fernando R. 2015. A fast and efficient Gibbs sampler for BayesB in whole-genome analyses. Genet Sel Evol, 47:80.\nCheng H, Kizilkaya K, Zeng J, Garrick D, Fernando R. 2018. Genomic Prediction from Multiple-Trait Bayesian Regression Methods Using Mixture Priors. Genetics, 209(1): 89-103.","category":"page"},{"location":"citing/citing/#Single-step-Methods","page":"Citing","title":"Single-step Methods","text":"","category":"section"},{"location":"citing/citing/","page":"Citing","title":"Citing","text":"Fernando R, Dekkers J, Garrick D. 2015. A class of Bayesian methods to combine large numbers of genotyped and non-genotyped animals for whole-genome analyses. Genetics Selection Evolution, 46(1), 50.\nFernando R, Cheng H, Golden B, Garrick D. 2016. Computational strategies for alternative single-step Bayesian regression models with large numbers of genotyped and non-genotyped animals. Genetics Selection Evolution, 48(1), 96.","category":"page"},{"location":"nnmm/Part1_introduction/#Part1.-introduction","page":"Part 1. Introduction","title":"Part1. introduction","text":"","category":"section"},{"location":"nnmm/Part1_introduction/#.-Overview","page":"Part 1. Introduction","title":"1. Overview","text":"","category":"section"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"The Mixed Effect Neural Networks (NN-MM) extend linear mixed model (\"MM\") to multilayer neural networks (\"NN\") by adding one middle layer between genotype layer and phenotypes layer. Nodes in the middle layer represent intermediate traits, e.g., the known intermediate omics features such as gene expression levels can be incorporated in the middle layer. These three sequential layers form a unified network. ","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"(Image: )","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"NN-MM allow any patterns of missing data in the middle layer, and missing data will be sampled. In below figure, for an individual, the gene expression levels of the first two genes are 0.9 and 0.1, respectively, and the gene expression level of the last gene is missing to be sampled. The missing patterns of gene expression levels can be different for different individuals.","category":"page"},{"location":"nnmm/Part1_introduction/#.-Extend-linear-mixed-model-to-multilayer-neural-networks","page":"Part 1. Introduction","title":"2. Extend linear mixed model to multilayer neural networks","text":"","category":"section"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"Multiple independent single-trait mixed models are used to model the relationships between input layer (genotypes) and middle layer (intermediate traits). Activation functions in the neural network are used to approximate the linear/nonlinear relationships between middle layer (intermediate traits) and output layer (phenotypes). Missing values in the middle layer (intermediate traits) are sampled by Hamiltonian Monte Carlo based on the upstream genotype layer and downstream phenotype layer.","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"Details can be found in our publications:","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"Tianjing Zhao, Jian Zeng, and Hao Cheng. Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data, GENETICS, 2022; https://doi.org/10.1093/genetics/iyac034. \nTianjing Zhao, Rohan Fernando, and Hao Cheng. Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies, G3 Genes|Genomes|Genetics, 2021;  https://doi.org/10.1093/g3journal/jkab228","category":"page"},{"location":"nnmm/Part1_introduction/#.-Flexibility","page":"Part 1. Introduction","title":"3. Flexibility","text":"","category":"section"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"NN-MM can fit fully-connected neural networks ((a),(b)), or partial-connected neural networks ((c),(d)). Also, the relationship between middle layer (intermediate traits) and output layer (phenotypes) can be based on activation functions ((a),(c)), or pre-defined by a user-defined function ((b),(d)).","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"(Image: )","category":"page"},{"location":"nnmm/Part1_introduction/#.-Multi-threaded-parallelism","page":"Part 1. Introduction","title":"4. Multi-threaded parallelism","text":"","category":"section"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"By default, multiple single-trait models will be used to model the relationships between input layer (genotypes) and middle layer (intermediate traits). Multi-threaded parallelism will be used for parallel computing. The number of threads can be checked by running Threads.nthreads() in Julia. Usually, using multiple threads will be about 3 times faster than using a single thread.","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"The number of execution threads is controlled by using the -t/–threads command-line argument (requires at least Julia 1.5). ","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"For example, to start Julia with 4 threads:","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"julia --threads 4","category":"page"},{"location":"nnmm/Part1_introduction/","page":"Part 1. Introduction","title":"Part 1. Introduction","text":"If you're using Juno via IDE like Atom, all threads will be loaded automatically. ","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Part-6.-NNMM-as-Traditional-Genomic-Prediction","page":"Part 6. Traditional genomic prediction","title":"Part 6. NNMM as Traditional Genomic Prediction","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Overview","page":"Part 6. Traditional genomic prediction","title":"Overview","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"NNMM is a general framework that encompasses traditional genomic prediction models as special cases. By configuring the network architecture appropriately, you can use NNMM to perform standard single-trait or multi-trait BayesC (and other Bayesian alphabet methods).","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"This demonstrates that NNMM generalizes traditional genomic prediction rather than replacing it.","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#The-Key-Insight","page":"Part 6. Traditional genomic prediction","title":"The Key Insight","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"Traditional genomic prediction models like BayesC can be expressed as a special case of NNMM by:","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"Using a single completely missing node in the middle layer - This latent node will be sampled based on genotype information\nUsing a linear activation function - This ensures the relationship between the latent node and phenotype is linear","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"┌─────────────┐     ┌─────────────────┐     ┌─────────────┐\n│  Genotypes  │ ──► │  Latent Node    │ ──► │ Phenotype   │\n│   (SNPs)    │     │ (100% missing)  │     │    (y)      │\n└─────────────┘     └─────────────────┘     └─────────────┘\n                    Sampled via BayesC       Linear activation","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"This is mathematically equivalent to the traditional model:","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"y = Xbeta + Zalpha + e","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"where alpha are marker effects sampled using BayesC.","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Example:-Single-Trait-BayesC","page":"Part 6. Traditional genomic prediction","title":"Example: Single-Trait BayesC","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"using NNMM, DataFrames, CSV\n\n# Step 1: Create a middle layer with ONE completely missing node\nomics_df = DataFrame(\n    ID = individual_ids,\n    latent = fill(missing, n_individuals)  # All NA\n)\nCSV.write(\"latent.csv\", omics_df; missingstring=\"NA\")\n\n# Step 2: Define 3-layer network\nlayers = [\n    Layer(layer_name=\"geno\", data_path=[\"genotypes.csv\"]),\n    Layer(layer_name=\"latent\", data_path=\"latent.csv\", missing_value=\"NA\"),\n    Layer(layer_name=\"phenotypes\", data_path=\"phenotypes.csv\", missing_value=\"NA\")\n]\n\n# Step 3: Define equations with LINEAR activation\nequations = [\n    # Genotypes → Latent (marker regression)\n    Equation(\n        from_layer_name=\"geno\",\n        to_layer_name=\"latent\",\n        equation=\"latent = intercept + geno\",\n        omics_name=[\"latent\"],\n        method=\"BayesC\",\n        estimatePi=true\n    ),\n    # Latent → Phenotype (LINEAR activation = standard regression)\n    Equation(\n        from_layer_name=\"latent\",\n        to_layer_name=\"phenotypes\",\n        equation=\"phenotypes = intercept + latent\",\n        phenotype_name=[\"y1\"],\n        method=\"BayesC\",\n        activation_function=\"linear\"  # Key: linear activation\n    )\n]\n\n# Step 4: Run analysis\nresult = runNNMM(layers, equations; chain_length=50000, burnin=10000)","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Example:-Multi-Trait-BayesC","page":"Part 6. Traditional genomic prediction","title":"Example: Multi-Trait BayesC","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"For multi-trait analysis, simply use multiple missing nodes:","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"# Middle layer with TWO completely missing nodes\nomics_df = DataFrame(\n    ID = individual_ids,\n    latent1 = fill(missing, n_individuals),\n    latent2 = fill(missing, n_individuals)\n)\nCSV.write(\"latent_multi.csv\", omics_df; missingstring=\"NA\")\n\n# Phenotype data with two traits\npheno_df = DataFrame(\n    ID = individual_ids,\n    y1 = trait1_values,\n    y2 = trait2_values\n)\n\nlayers = [\n    Layer(layer_name=\"geno\", data_path=[\"genotypes.csv\"]),\n    Layer(layer_name=\"latent\", data_path=\"latent_multi.csv\", missing_value=\"NA\"),\n    Layer(layer_name=\"phenotypes\", data_path=\"phenotypes_multi.csv\", missing_value=\"NA\")\n]\n\nequations = [\n    Equation(\n        from_layer_name=\"geno\",\n        to_layer_name=\"latent\",\n        equation=\"latent = intercept + geno\",\n        omics_name=[\"latent1\", \"latent2\"],\n        method=\"BayesC\"\n    ),\n    Equation(\n        from_layer_name=\"latent\",\n        to_layer_name=\"phenotypes\",\n        equation=\"phenotypes = intercept + latent\",\n        phenotype_name=[\"y1\", \"y2\"],\n        method=\"BayesC\",\n        activation_function=\"linear\"\n    )\n]\n\nresult = runNNMM(layers, equations; chain_length=50000)","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Comparison-with-Direct-BayesC","page":"Part 6. Traditional genomic prediction","title":"Comparison with Direct BayesC","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"Aspect Traditional BayesC NNMM as BayesC\nModel equation y = intercept + geno 3-layer network with missing middle\nMarker effects Directly estimated Estimated via latent layer\nFlexibility Fixed model Can add intermediate omics later\nResults EBV, marker effects EBV, marker effects","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Why-Use-NNMM-for-Traditional-Analysis?","page":"Part 6. Traditional genomic prediction","title":"Why Use NNMM for Traditional Analysis?","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"Unified framework: Same code structure for simple and complex models\nEasy extension: Add observed omics data to the middle layer when available\nGradual complexity: Start with traditional model, add neural network components as needed\nFuture-proof: As more omics data becomes available, simply fill in the missing values","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Other-Bayesian-Methods","page":"Part 6. Traditional genomic prediction","title":"Other Bayesian Methods","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"You can use any Bayesian alphabet method by changing the method parameter:","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"# BayesA\nEquation(..., method=\"BayesA\")\n\n# BayesB  \nEquation(..., method=\"BayesB\", estimatePi=true)\n\n# BayesL (Bayesian LASSO)\nEquation(..., method=\"BayesL\")\n\n# RR-BLUP (Ridge Regression)\nEquation(..., method=\"RR-BLUP\")","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/#Summary","page":"Part 6. Traditional genomic prediction","title":"Summary","text":"","category":"section"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"NNMM provides a superset of traditional genomic prediction methods:","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"Traditional BayesC = NNMM with completely missing middle layer + linear activation\nNNMM with omics = Same framework with observed intermediate traits\nFull neural network = Same framework with nonlinear activation functions","category":"page"},{"location":"nnmm/Part6_TraditionalGenomicPrediction/","page":"Part 6. Traditional genomic prediction","title":"Part 6. Traditional genomic prediction","text":"This design allows researchers to start with familiar methods and gradually incorporate more complex models as data and understanding grow.","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Mixed-effect-neural-network:-Genotypes-(complete/incomplete)-Intemediate-omics-features-Phenotyes","page":"Part 3. NNMM with intermediate omics","title":"Mixed effect neural network: Genotypes -> (complete/incomplete) Intemediate omics features -> Phenotyes","text":"","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"Tips:","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"Put the names of omics features in the build_model() function through the latent_traits argument.\nIf there are many omics features (e.g., 1000), it is recommanded to set printout_model_info=false in the runMCMC() function.\nMissing omics data for individuals in the training dataset (i.e., individuals with phenotypes) is allowed. When you read a file with missing values via the CSV.read() function, the missingstrings argument should be used to set sentinel values that will be parsed as missing.\nFor individuals in the testing dataset (i.e., individuals without phenotypes), if the testing individuals have complete omics data, then incorporating the omics data of those individuals may improve the relationship between input layer (genotype) and middle layer (omics).","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#example(o1):-fully-connected-neural-networks-with-observed-intemediate-omics-features","page":"Part 3. NNMM with intermediate omics","title":"example(o1): fully-connected neural networks with observed intemediate omics features","text":"","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"nonlinear function (to define relationship between omics and phenotye): sigmoid (other supported activation functions: \"tanh\", \"relu\", \"leakyrelu\", \"linear\")\nnumber of omics features in the middle layer: 10\nBayesian model: multiple independent single-trait BayesC (to sample marker effects on intemediate omics)\nsample the missing omics in the middle layer: Hamiltonian Monte Carlo","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"(Image: )","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"# Step 1: Load packages\r\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets, Random, HTTP #HTTP to download demo data from github\r\nRandom.seed!(123)\r\n\r\n# Step 2: Read data (from github)\r\nphenofile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/y.csv\").body\r\nomicsfile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/omics.csv\").body\r\ngenofile   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_n100_p200.csv\").body\r\nphenotypes = CSV.read(phenofile,DataFrame)\r\nomics      = CSV.read(omicsfile,DataFrame)\r\ngeno_df    = CSV.read(genofile,DataFrame)\r\n\r\nomics_names = names(omics)[2:end]  #get names of omics\r\ninsertcols!(omics,2,:y => phenotypes[:,:y], :bv => phenotypes[:,:bv]) #phenotype and omics should be in the same dataframe\r\n\r\ngenotypes = get_genotypes(geno_df,separator=',',method=\"BayesC\")\r\n\r\n# Step 3: Build Model Equations\r\nmodel_equation  =\"y = intercept + genotypes\" #name of phenotypes is \"y\"\r\n                                             #name of genotypes is \"genotypes\" (user-defined in the previous step)\r\n                                             #the single-trait mixed model used between input and each omics is: omics = intercept + genotypes\r\nmodel = build_model(model_equation,\r\n\t\t    num_hidden_nodes=10,          #number of omics in middle layer is 3\r\n                    latent_traits=omics_names,    #name of all omics\r\n\t\t    nonlinear_function=\"sigmoid\") #sigmoid function is used to approximate relationship between omics and phenotypes\r\n\r\n# Step 4: Run Analysis\r\nout=runMCMC(model, omics, chain_length=5000, printout_model_info=false);\r\n\r\n# Step 5: Check Accuruacy\r\nresults    = innerjoin(out[\"EBV_NonLinear\"], omics, on = :ID)\r\naccuruacy  = cor(results[!,:EBV],results[!,:bv])","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Includes-a-residual-that-is-not-mediated-by-other-omics-features","page":"Part 3. NNMM with intermediate omics","title":"Includes a residual that is not mediated by other omics features","text":"","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"To include residuals polygenic component (i.e. directly from genotypes to phenotypes, not mediated by omics features), you can additional hidden nodes in the middle layer (see example (o2)). This can also be achieved in a partial-connected neural network in a same manner.","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"(Image: )","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#example(o2):-fully-connected-neural-network-with-residuals","page":"Part 3. NNMM with intermediate omics","title":"example(o2): fully-connected neural network with residuals","text":"","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"For all individuals, this extra hidden node will be treated as unknown to be sampled.","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"# Step 1: Load packages\r\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets, Random, HTTP \r\nRandom.seed!(123)\r\n\r\n# Step 2: Read data (from github)\r\nphenofile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/y.csv\").body\r\nomicsfile  = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/omics.csv\").body\r\ngenofile   = HTTP.get(\"https://raw.githubusercontent.com/zhaotianjing/nnmm_doc/main/data_simulation/geno_n100_p200.csv\").body\r\nphenotypes = CSV.read(phenofile,DataFrame)\r\nomics      = CSV.read(omicsfile,DataFrame)\r\ngeno_df    = CSV.read(genofile,DataFrame)\r\n\r\ninsertcols!(omics, :residual => missing)  #create a hidden node to account for residuals\r\nomics[!,:residual] = convert(Vector{Union{Missing,Float64}}, omics[!,:residual]) #transform the datatype is required for Julia\r\nomics_names = names(omics)[2:end]  #get names of 10 omics and 1 hidden node\r\ninsertcols!(omics,2,:y => phenotypes[:,:y], :bv => phenotypes[:,:bv]) #phenotype and omics should be in the same dataframe\r\n\r\ngenotypes = get_genotypes(geno_df,separator=',',method=\"BayesC\")\r\n\r\n# Step 3: Build Model Equations\r\nmodel_equation  =\"y = intercept + genotypes\" \r\nmodel = build_model(model_equation,\r\n\t\t    num_hidden_nodes=11,   #10 omcis and 1 hidden node\r\n                    latent_traits=omics_names,\r\n\t\t    nonlinear_function=\"sigmoid\")\r\n\r\n# Step 4: Run Analysis\r\nout = runMCMC(model,omics,chain_length=5000,printout_model_info=false)\r\n\r\n# Step 5: Check Accuruacy\r\nresults    = innerjoin(out[\"EBV_NonLinear\"], omics, on = :ID)\r\naccuruacy  = cor(results[!,:EBV],results[!,:bv])","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"Users can also add extra hidden nodes in the partial-connected neural network. Please check next documentation for building a partial-connected neural network.","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Output-files","page":"Part 3. NNMM with intermediate omics","title":"Output files","text":"","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"Same as those described in Part2.","category":"page"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/#Julia-Tips:","page":"Part 3. NNMM with intermediate omics","title":"Julia Tips:","text":"","category":"section"},{"location":"nnmm/Part3_NNMMwithIntermediateOmics/","page":"Part 3. NNMM with intermediate omics","title":"Part 3. NNMM with intermediate omics","text":"You may want to set missing values manually, for example, setting the phenotypes for individuals in testing dataset as missing. Firstly, the type of  columns should be changed to allow missing, e.g., phenotypes[!,:y] =  convert(Vector{Union{Missing,Float64}}, phenotypes[!,:y]). Then, missing can be set manually, e.g., phenotypes[10:11,:y1] .= missing forces the 10th and 11th elements to be missing.","category":"page"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/#User-defined-relationship-between-middle-layer-(intermediate-traits)-and-output-layer-(phenotype)","page":"Part 5. User-defined nonlinear function","title":"User-defined relationship between middle layer (intermediate traits) and output layer (phenotype)","text":"","category":"section"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/","page":"Part 5. User-defined nonlinear function","title":"Part 5. User-defined nonlinear function","text":"Firstly, a function should be pre-defined by user, where each function argument represents the input from a intermediate trait. For example, with two omics, we can have pig_growth(omics1,omics2) = sqrt(omics1^2 / (omics1^2 + omics2^2)))\nThen, put the name of the user-define function in the nonlinear_function arugument when build the model.","category":"page"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/#example(b):-user-defined-relationship-between-middle-layer-(intermediate-traits)-and-output-layer-(phenotype)","page":"Part 5. User-defined nonlinear function","title":"example(b): user-defined relationship between middle layer (intermediate traits) and output layer (phenotype)","text":"","category":"section"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/","page":"Part 5. User-defined nonlinear function","title":"Part 5. User-defined nonlinear function","text":"number of nodes in the middle layer: 2\nnonlinear function (to define relationship between middle nodes and phenotype): y = sqrt(x1^2 / (x1^2 + x2^2))\nsample the missing omics in the middle layer: Matropolis-Hastings","category":"page"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/","page":"Part 5. User-defined nonlinear function","title":"Part 5. User-defined nonlinear function","text":"All the other code are the same as before, except Step 3: Build Model Equations. Note that user-defined relationship is supported in both fully-connected and partial-connected neural network. ","category":"page"},{"location":"nnmm/Part5_UserDefinedNonlinearFunction/","page":"Part 5. User-defined nonlinear function","title":"Part 5. User-defined nonlinear function","text":"# Step 3: Build Model Equations\r\npig_growth(x1,x2) = sqrt(x1^2 / (x1^2 + x2^2))\r\nmodel = build_model(model_equation,\r\n                    num_hidden_nodes=2,\r\n                    nonlinear_function=pig_growth);","category":"page"},{"location":"manual/getstarted/#Get-Started","page":"Get Started","title":"Get Started","text":"","category":"section"},{"location":"manual/getstarted/#Installation","page":"Get Started","title":"Installation","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To install Julia, please go to the official Julia website. Please see platform specific instructions if you have trouble installing Julia.","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To install the NNMM package, use the following command inside the Julia REPL (or IJulia Notebook):","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"using Pkg\nPkg.add(\"NNMM\")","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To load the NNMM package:","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"using NNMM","category":"page"},{"location":"manual/getstarted/#Development-Version","page":"Get Started","title":"Development Version","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To use the latest/beta features under development:","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"Pkg.add(url=\"https://github.com/reworkhow/NNMM.jl\")","category":"page"},{"location":"manual/getstarted/#Jupyter-Notebook","page":"Get Started","title":"Jupyter Notebook","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"If you prefer \"reproducible research\", an interactive Jupyter Notebook interface is available for Julia (and therefore NNMM). The Jupyter Notebook is an open-source web application for creating and sharing documents that contain live code, equations, visualizations and explanatory text. To install IJulia for Jupyter Notebook, please go to IJulia.","category":"page"},{"location":"manual/getstarted/#Multi-threaded-Parallelism","page":"Get Started","title":"Multi-threaded Parallelism","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"NNMM supports multi-threaded parallelism for faster computation. The number of threads can be checked by:","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"Threads.nthreads()","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To start Julia with multiple threads (requires Julia 1.5+):","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"julia --threads 4","category":"page"},{"location":"manual/getstarted/#Access-Documentation","page":"Get Started","title":"Access Documentation","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"warning: Warning\nPlease load the NNMM package first.","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To show basic information about NNMM in REPL or IJulia notebook, use ?NNMM and press enter.","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"For help on a specific function, type ? followed by its name, e.g., ?runNNMM and press enter.","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"The full documentation is available here.","category":"page"},{"location":"manual/getstarted/#Run-Your-Analysis","page":"Get Started","title":"Run Your Analysis","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"There are several ways to run your analysis:","category":"page"},{"location":"manual/getstarted/#Interactive-Session-(REPL)","page":"Get Started","title":"Interactive Session (REPL)","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"Start an interactive session by double-clicking the Julia executable or running julia from the command line:","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"julia> using NNMM\njulia> # your analysis code here","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To evaluate code written in a file script.jl in REPL:","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"julia> include(\"script.jl\")","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To exit the interactive session, type ^D (control + d) or quit().","category":"page"},{"location":"manual/getstarted/#Command-Line","page":"Get Started","title":"Command Line","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To run code in a file non-interactively from the command line:","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"julia script.jl","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"If you want to pass arguments to your script:","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"julia script.jl arg1 arg2","category":"page"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"where arguments arg1 and arg2 are passed as ARGS[1] and ARGS[2] of type String.","category":"page"},{"location":"manual/getstarted/#Jupyter-Notebook-2","page":"Get Started","title":"Jupyter Notebook","text":"","category":"section"},{"location":"manual/getstarted/","page":"Get Started","title":"Get Started","text":"To run code in Jupyter Notebook, please see IJulia.","category":"page"},{"location":"nnmm/Part2_NNMM/#Mixed-effect-neural-network:-Genotypes-Unobserved-intemediate-traits-Phenotyes","page":"Part 2. NNMM","title":"Mixed effect neural network: Genotypes -> Unobserved intemediate traits -> Phenotyes","text":"","category":"section"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"Tips: please center the phenotypes to have zero mean.","category":"page"},{"location":"nnmm/Part2_NNMM/#Example(a):-fully-connected-neural-networks,-all-intemediate-traits-are-unobserved","page":"Part 2. NNMM","title":"Example(a): fully-connected neural networks, all intemediate traits are unobserved","text":"","category":"section"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"nonlinear function (to define relationship between middle layer and phenotye): tanh (other supported activation functions: \"sigmoid\", \"relu\", \"leakyrelu\", \"linear\")\nnumber of nodes in the middle layer: 3\nBayesian model: multiple independent single-trait BayesC (to sample marker effects on intemediate traits). Note, to use multi-trait Bayesian Alphabet models, please set mega_trait=false in runMCMC() function.\nsample the unobserved intemediate traits in the middle layer: Hamiltonian Monte Carlo","category":"page"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"(Image: )","category":"page"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"# Step 1: Load packages\r\nusing NNMM,DataFrames,CSV,Statistics,NNMM.Datasets,Random\r\nRandom.seed!(123)\r\n\r\n# Step 2: Read data \r\nphenofile  = dataset(\"phenotypes.csv\") #get example data path\r\ngenofile   = dataset(\"genotypes.csv\")  #get example data path\r\n\r\nphenotypes = CSV.read(phenofile,DataFrame,delim = ',',header=true,missingstrings=[\"NA\"]) #read phenotypes (output layer)\r\ngenotypes  = get_genotypes(genofile,separator=',',method=\"BayesC\");                      #read genotypes  (input layer)\r\n\r\n\r\n# Step 3: Build Model Equations \r\nmodel_equation  =\"y1 = intercept + genotypes\"  #name of phenotypes is \"y1\" in the phenotypes data\r\n                                               #name of genotypes is \"genotypes\" (user-defined in the previous step)\r\n                                               #the single-trait mixed model used between input and each node in middle layer is: middle node = intercept + genotypes\r\nmodel = build_model(model_equation,\r\n\t\t    num_hidden_nodes=3,            #number of nodes in middle layer is 3\r\n\t\t    nonlinear_function=\"tanh\");    #tanh function is used to approximate relationship between middle layer and phenotype\r\n\r\n\r\n# Step 4: Run Analysis\r\nout=runMCMC(model,phenotypes,chain_length=5000); \r\n\r\n# Step 5: Check Accuruacy\r\nresults    = innerjoin(out[\"EBV_NonLinear\"], phenotypes, on = :ID) \r\naccuruacy  = cor(results[!,:EBV],results[!,:bv1])","category":"page"},{"location":"nnmm/Part2_NNMM/#Example-output-files","page":"Part 2. NNMM","title":"Example output files","text":"","category":"section"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"The i-th middle nodes will be named as \"trait name\"+\"i\". In our example, the observed trait is named \"y1\", and there are 3 middle nodes, so the middle nodes are named as \"y11\", \"y12\", and \"y13\", respectively.","category":"page"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"Below is a list of files containing estimates and standard deviations for variables of interest. ","category":"page"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"file name description\nEBV_NonLinear.txt estimated breeding values for observed trait\nEBV_y11.txt estimated breeding values for middle node 1\nEBV_y12.txt estimated breeding values for middle node 2\nEBV_y13.txt estimated breeding values for middle node 3\ngenetic_variance.txt estimated genetic variance-covariance of all middle nodes\nheritability.txt estimated heritability of all middle nodes\nlocation_parameters.txt estimated bias of all middle nodes\nneuralnetworksbiasandweights.txt. estimated bias of phenotypes and weights between middle nodes and phenotypes\npi_genotypes.txt estimated pi of all middle nodes\nmarkereffectsgenotypes.txt estimated marker effects of all middle nodes\nresidual_variance.txt estimated residual variance-covariance for all middle nodes","category":"page"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"Below is a list of files containing MCMC samples for variables of interest. ","category":"page"},{"location":"nnmm/Part2_NNMM/","page":"Part 2. NNMM","title":"Part 2. NNMM","text":"file name description\nMCMCsamplesEBV_NonLinear.txt MCMC samples from the posterior distribution of breeding values for phenotypes\nMCMCsamplesEBV_y11.txt MCMC samples from the posterior distribution of breeding values for middle node 1\nMCMCsamplesEBV_y12.txt MCMC samples from the posterior distribution of breeding values for middle node 2\nMCMCsamplesEBV_y13.txt MCMC samples from the posterior distribution of breeding values for middle node 3\nMCMCsamplesgenetic_variance.txt MCMC samples from the posterior distribution of genetic variance-covariance for all middle nodes\nMCMCsamplesheritability.txt MCMC samples from the posterior distribution of heritability for all middle nodes\nMCMCsamplesmarkereffectsgenotypes_y11 MCMC samples from the posterior distribution of marker effects for middle node 1\nMCMCsamplesmarkereffectsgenotypes_y12 MCMC samples from the posterior distribution of marker effects for middle node 2\nMCMCsamplesmarkereffectsgenotypes_y13 MCMC samples from the posterior distribution of marker effects for middle node 3\nMCMCsamplesmarkereffectsvariances_genotypes.txt MCMC samples from the posterior distribution of marker effect variance for all middle nodes\nMCMCsamplesneuralnetworksbiasandweights.txt. MCMC samples from the posterior distribution of bias of observed trait and weights between middle nodes and phenotypes\nMCMCsamplespi_genotypes.txt MCMC samples from the posterior distribution of pi for all middle nodes\nMCMCsamplesresidual_variance.txt MCMC samples from the posterior distribution of residual variance-covariance for all middle nodes","category":"page"},{"location":"","page":"Home","title":"Home","text":"(Image: NNMM)","category":"page"},{"location":"#NNMM.jl-Neural-Network-Mixed-Models","page":"Home","title":"NNMM.jl - Neural Network Mixed Models","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"NNMM.jl is a Julia package for Mixed Effect Neural Networks that extend traditional linear mixed models to multilayer neural networks for genomic prediction and genome-wide association studies.","category":"page"},{"location":"#Key-Features","page":"Home","title":"Key Features","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Generalizes Traditional Methods: Traditional BayesC and other Bayesian alphabet methods are special cases of NNMM\nNeural Network Architecture: Extend linear mixed models to multilayer neural networks\nIntermediate Omics Integration: Incorporate known intermediate omics features (e.g., gene expression) in the middle layer\nFlexible Missing Data Handling: Allow any patterns of missing data in intermediate layers\nBayesian Framework: Full Bayesian inference using MCMC and Hamiltonian Monte Carlo\nMulti-threaded Parallelism: Efficient parallel computing for large-scale analyses","category":"page"},{"location":"#Network-Architecture","page":"Home","title":"Network Architecture","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"NNMM models relationships between:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Input Layer: Genotypes (SNP markers)\nMiddle Layer: Intermediate traits (e.g., gene expression, metabolomics)\nOutput Layer: Phenotypes","category":"page"},{"location":"","page":"Home","title":"Home","text":"The package supports:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Fully-connected neural networks\nPartial-connected neural networks\nUser-defined nonlinear activation functions","category":"page"},{"location":"#Installation","page":"Home","title":"Installation","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"using Pkg\nPkg.add(\"NNMM\")","category":"page"},{"location":"","page":"Home","title":"Home","text":"Or for the latest development version:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Pkg.add(url=\"https://github.com/reworkhow/NNMM.jl\")","category":"page"},{"location":"#Quick-Start","page":"Home","title":"Quick Start","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"using NNMM, CSV, DataFrames\n\n# Load data\nphenotypes = CSV.read(\"phenotypes.csv\", DataFrame)\nomics = CSV.read(\"omics.csv\", DataFrame)\ngenotypes = CSV.read(\"genotypes.csv\", DataFrame)\n\n# Define network layers\nlayer1 = Layer(data=genotypes)\nlayer2 = Layer(data=omics, n_latent=size(omics, 2)-1)\nlayer3 = Layer(data=phenotypes)\n\n# Build model equation\neq = Equation(layer1, layer2, layer3, \"y\")\n\n# Run MCMC\nresults = runNNMM(eq, chain_length=5000, output_folder=\"results\")","category":"page"},{"location":"#Supporting-and-Citing","page":"Home","title":"Supporting and Citing","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"If you use NNMM.jl for your research, please cite:","category":"page"},{"location":"","page":"Home","title":"Home","text":"Zhao, T., Zeng, J., & Cheng, H. (2022). Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. GENETICS. https://doi.org/10.1093/genetics/iyac034","category":"page"},{"location":"","page":"Home","title":"Home","text":"Zhao, T., Fernando, R., & Cheng, H. (2021). Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. G3 Genes|Genomes|Genetics. https://doi.org/10.1093/g3journal/jkab228","category":"page"},{"location":"","page":"Home","title":"Home","text":"Please star the repository here to help demonstrate community involvement.","category":"page"},{"location":"#Getting-Help","page":"Home","title":"Getting Help","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Open an issue on GitHub\nContact: <qtlcheng@ucdavis.edu>","category":"page"},{"location":"#Tutorials","page":"Home","title":"Tutorials","text":"","category":"section"},{"location":"#NNMM-Tutorials","page":"Home","title":"NNMM Tutorials","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Pages = [\n  \"nnmm/Part1_introduction.md\",\n  \"nnmm/Part2_NNMM.md\",\n  \"nnmm/Part3_NNMMwithIntermediateOmics.md\",\n  \"nnmm/Part4_PartialConnectedNeuralNetwork.md\",\n  \"nnmm/Part5_UserDefinedNonlinearFunction.md\",\n  \"nnmm/Part6_TraditionalGenomicPrediction.md\",\n]\nDepth = 2","category":"page"},{"location":"#Manual","page":"Home","title":"Manual","text":"","category":"section"},{"location":"","page":"Home","title":"Home","text":"Pages = [\n  \"manual/getstarted.md\",\n  \"manual/public.md\",\n]\nDepth = 2","category":"page"},{"location":"manual/public/#Public-Functions","page":"Public Functions","title":"Public Functions","text":"","category":"section"},{"location":"manual/public/","page":"Public Functions","title":"Public Functions","text":"Documentation for NNMM.jl's public interface. Below are functions available to general users.","category":"page"},{"location":"manual/public/#Index","page":"Public Functions","title":"Index","text":"","category":"section"},{"location":"manual/public/","page":"Public Functions","title":"Public Functions","text":"Pages = [\"public.md\"]\nModules = [NNMM]","category":"page"},{"location":"manual/public/#NNMM-Functions","page":"Public Functions","title":"NNMM Functions","text":"","category":"section"},{"location":"manual/public/#Core-Types","page":"Public Functions","title":"Core Types","text":"","category":"section"},{"location":"manual/public/","page":"Public Functions","title":"Public Functions","text":"Layer\nEquation","category":"page"},{"location":"manual/public/#NNMM.Layer","page":"Public Functions","title":"NNMM.Layer","text":"Layer\n\nA struct representing a layer in the Neural Network Mixed Model (NNMM).\n\nFields\n\nlayer_name::String: Name identifier for the layer\ndata_path: Path to the data file(s) - String for single file, Vector{String} for partial connection\nseparator: Column separator in the data file (default: ',')\nheader: Whether the data file has a header row (default: true)\ndata: Loaded data (initialized as empty)\nquality_control: Whether to perform quality control on genotypes (default: true)\nMAF: Minor Allele Frequency threshold for QC (default: 0.01)\nmissing_value: Value representing missing data (default: 9.0)\ncenter: Whether to center the data (default: true)\n\n\n\n\n\n","category":"type"},{"location":"manual/public/#NNMM.Equation","page":"Public Functions","title":"NNMM.Equation","text":"Equation\n\nA struct representing an equation connecting layers in NNMM.\n\nFields\n\nfrom_layer_name: Source layer name\nto_layer_name: Target layer name\nequation: Model equation string\nomics_name: Names of omics variables (or false if not applicable)\nphenotype_name: Names of phenotype variables (or false if not applicable)\ncovariate: Covariate variable names\nrandom: Random effect specifications\nactivation_function: Activation function (\"linear\", \"sigmoid\", \"tanh\", etc.)\npartial_connect_structure: Structure for partial connectivity\nstarting_value: Starting values for MCMC\nmethod: Bayesian method (\"BayesC\", \"BayesA\", \"BayesB\", etc.)\nPi: Prior probability of zero effect\nestimatePi: Whether to estimate Pi\nVariance parameters for G (genetic) and R (residual)\n\n\n\n\n\n","category":"type"},{"location":"manual/public/#Main-Functions","page":"Public Functions","title":"Main Functions","text":"","category":"section"},{"location":"manual/public/","page":"Public Functions","title":"Public Functions","text":"runNNMM","category":"page"},{"location":"manual/public/#Utility-Functions","page":"Public Functions","title":"Utility Functions","text":"","category":"section"},{"location":"manual/public/","page":"Public Functions","title":"Public Functions","text":"get_pedigree\nget_genotypes","category":"page"},{"location":"manual/public/#NNMM.PedModule.get_pedigree","page":"Public Functions","title":"NNMM.PedModule.get_pedigree","text":"get_pedigree(pedfile::AbstractString;header=false,separator=',',missingstrings=[\"0\"])\n\nGet pedigree informtion from a pedigree file with header (defaulting to false) , separator (defaulting to ,) and missing values (defaulting to [\"0\"])\nPedigree file format:\n\na,0,0\nc,a,b\nd,a,c\n\n\n\n\n\n","category":"function"}]
}
