# Citing

## Software

If you use NNMM.jl, please cite:

* Zhao, T., Zeng, J., & Cheng, H. (2022). Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. *GENETICS*, iyac034. https://doi.org/10.1093/genetics/iyac034

* Zhao, T., Fernando, R., & Cheng, H. (2021). Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. *G3 Genes|Genomes|Genetics*, jkab228. https://doi.org/10.1093/g3journal/jkab228

## Methods

### Neural Network Mixed Models (NNMM)

* Zhao T, Zeng J, Cheng H. 2022. Extend mixed models to multilayer neural networks for genomic prediction including intermediate omics data. GENETICS, iyac034.

* Zhao T, Fernando R, Cheng H. 2021. Interpretable artificial neural networks incorporating Bayesian alphabet models for genome-wide prediction and association studies. G3 Genes|Genomes|Genetics, jkab228.

### Bayesian Alphabet Methods

* Meuwissen T, Hayes B, Goddard M. 2001. Prediction of total genetic value using genome-wide dense marker maps. Genetics, 157:1819â€“1829.

* Habier D, Fernando R, Kizilkaya K, Garrick D. 2011. Extension of the bayesian alphabet for genomic selection. BMC Bioinformatics, 12(1), 186.

* Cheng H, Garrick D, Fernando R. 2015. A fast and efficient Gibbs sampler for BayesB in whole-genome analyses. Genet Sel Evol, 47:80.

* Cheng H, Kizilkaya K, Zeng J, Garrick D, Fernando R. 2018. Genomic Prediction from Multiple-Trait Bayesian Regression Methods Using Mixture Priors. Genetics, 209(1): 89-103.

### Single-step Methods

* Fernando R, Dekkers J, Garrick D. 2015. A class of Bayesian methods to combine large numbers of genotyped and non-genotyped animals for whole-genome analyses. Genetics Selection Evolution, 46(1), 50.

* Fernando R, Cheng H, Golden B, Garrick D. 2016. Computational strategies for alternative single-step Bayesian regression models with large numbers of genotyped and non-genotyped animals. Genetics Selection Evolution, 48(1), 96.
